<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.8.25">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>HPDM208Z Stratified Medicine Workshop 8: Genetic Association Analysis of Common Human Traits</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="hpdm208z_workshop8_files/libs/clipboard/clipboard.min.js"></script>
<script src="hpdm208z_workshop8_files/libs/quarto-html/quarto.js" type="module"></script>
<script src="hpdm208z_workshop8_files/libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="hpdm208z_workshop8_files/libs/quarto-html/axe/axe-check.js" type="module"></script>
<script src="hpdm208z_workshop8_files/libs/quarto-html/popper.min.js"></script>
<script src="hpdm208z_workshop8_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="hpdm208z_workshop8_files/libs/quarto-html/anchor.min.js"></script>
<link href="hpdm208z_workshop8_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="hpdm208z_workshop8_files/libs/quarto-html/quarto-syntax-highlighting-7b89279ff1a6dce999919e0e67d4d9ec.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="hpdm208z_workshop8_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="hpdm208z_workshop8_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="hpdm208z_workshop8_files/libs/bootstrap/bootstrap-d5214d16172734abcd1dcccc621ce81b.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">


</head>

<body class="quarto-light">

<header id="title-block-header" class="quarto-title-block default page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <h1 class="title">HPDM208Z Stratified Medicine Workshop 8: Genetic Association Analysis of Common Human Traits</h1>
            <p class="subtitle lead">MSc Health Data Science, University of Exeter</p>
                      </div>
  </div>
    
  
  <div class="quarto-title-meta">

      
    
      
    </div>
    
  
  </header><div id="quarto-content" class="page-columns page-rows-contents page-layout-article">
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
  <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction"><span class="header-section-number">1</span> Introduction</a></li>
  <li><a href="#aims-of-workshop" id="toc-aims-of-workshop" class="nav-link" data-scroll-target="#aims-of-workshop"><span class="header-section-number">2</span> Aims of workshop</a></li>
  <li><a href="#conda-environment" id="toc-conda-environment" class="nav-link" data-scroll-target="#conda-environment"><span class="header-section-number">3</span> Conda environment</a></li>
  <li><a href="#workshop-directory" id="toc-workshop-directory" class="nav-link" data-scroll-target="#workshop-directory"><span class="header-section-number">4</span> Workshop directory</a></li>
  <li><a href="#workshop-data-and-files" id="toc-workshop-data-and-files" class="nav-link" data-scroll-target="#workshop-data-and-files"><span class="header-section-number">5</span> Workshop data and files</a></li>
  <li><a href="#running-a-gwas-for-adult-human-height" id="toc-running-a-gwas-for-adult-human-height" class="nav-link" data-scroll-target="#running-a-gwas-for-adult-human-height"><span class="header-section-number">6</span> Running a GWAS for adult human height</a>
  <ul class="collapse">
  <li><a href="#task" id="toc-task" class="nav-link" data-scroll-target="#task"><span class="header-section-number">6.1</span> Task</a></li>
  </ul></li>
  <li><a href="#preparing-results-for-meta-analysis" id="toc-preparing-results-for-meta-analysis" class="nav-link" data-scroll-target="#preparing-results-for-meta-analysis"><span class="header-section-number">7</span> Preparing results for meta-analysis</a>
  <ul class="collapse">
  <li><a href="#ensuring-we-know-which-is-the-effect-allele" id="toc-ensuring-we-know-which-is-the-effect-allele" class="nav-link" data-scroll-target="#ensuring-we-know-which-is-the-effect-allele"><span class="header-section-number">7.1</span> Ensuring we know which is the effect allele</a></li>
  <li><a href="#obtaining-the-standard-error-of-the-effect-estimate" id="toc-obtaining-the-standard-error-of-the-effect-estimate" class="nav-link" data-scroll-target="#obtaining-the-standard-error-of-the-effect-estimate"><span class="header-section-number">7.2</span> Obtaining the standard error of the effect estimate</a></li>
  </ul></li>
  <li><a href="#meta-analysis-of-the-gwas-data" id="toc-meta-analysis-of-the-gwas-data" class="nav-link" data-scroll-target="#meta-analysis-of-the-gwas-data"><span class="header-section-number">8</span> Meta-analysis of the GWAS data</a>
  <ul class="collapse">
  <li><a href="#question" id="toc-question" class="nav-link" data-scroll-target="#question"><span class="header-section-number">8.1</span> Question</a></li>
  </ul></li>
  <li><a href="#dealing-with-population-stratification" id="toc-dealing-with-population-stratification" class="nav-link" data-scroll-target="#dealing-with-population-stratification"><span class="header-section-number">9</span> Dealing with population stratification</a>
  <ul class="collapse">
  <li><a href="#football-world-cup" id="toc-football-world-cup" class="nav-link" data-scroll-target="#football-world-cup"><span class="header-section-number">9.1</span> Football World Cup</a></li>
  <li><a href="#task---gwas-of-east-asian-football-team-support" id="toc-task---gwas-of-east-asian-football-team-support" class="nav-link" data-scroll-target="#task---gwas-of-east-asian-football-team-support"><span class="header-section-number">9.2</span> Task - GWAS of East-Asian football team support</a></li>
  <li><a href="#labelling-and-quantifying-differences-in-genetic-ancestry" id="toc-labelling-and-quantifying-differences-in-genetic-ancestry" class="nav-link" data-scroll-target="#labelling-and-quantifying-differences-in-genetic-ancestry"><span class="header-section-number">9.3</span> Labelling and quantifying differences in genetic ancestry</a></li>
  <li><a href="#performing-pca-on-the-reference-dataset" id="toc-performing-pca-on-the-reference-dataset" class="nav-link" data-scroll-target="#performing-pca-on-the-reference-dataset"><span class="header-section-number">9.4</span> Performing PCA on the reference dataset</a>
  <ul class="collapse">
  <li><a href="#task-1" id="toc-task-1" class="nav-link" data-scroll-target="#task-1"><span class="header-section-number">9.4.1</span> Task</a></li>
  </ul></li>
  <li><a href="#projecting-study-samples-into-the-reference-pc-space" id="toc-projecting-study-samples-into-the-reference-pc-space" class="nav-link" data-scroll-target="#projecting-study-samples-into-the-reference-pc-space"><span class="header-section-number">9.5</span> Projecting study samples into the reference PC space</a></li>
  <li><a href="#adjusting-association-analysis-for-pcs" id="toc-adjusting-association-analysis-for-pcs" class="nav-link" data-scroll-target="#adjusting-association-analysis-for-pcs"><span class="header-section-number">9.6</span> Adjusting association analysis for PCs</a></li>
  </ul></li>
  <li><a href="#classifying-genetic-ancestry" id="toc-classifying-genetic-ancestry" class="nav-link" data-scroll-target="#classifying-genetic-ancestry"><span class="header-section-number">10</span> Classifying genetic ancestry</a>
  <ul class="collapse">
  <li><a href="#defining-targets" id="toc-defining-targets" class="nav-link" data-scroll-target="#defining-targets"><span class="header-section-number">10.1</span> Defining targets</a></li>
  <li><a href="#creating-and-using-a-grid-search" id="toc-creating-and-using-a-grid-search" class="nav-link" data-scroll-target="#creating-and-using-a-grid-search"><span class="header-section-number">10.2</span> Creating and using a grid search</a></li>
  <li><a href="#assessing-performance-based-on-cross-validation" id="toc-assessing-performance-based-on-cross-validation" class="nav-link" data-scroll-target="#assessing-performance-based-on-cross-validation"><span class="header-section-number">10.3</span> Assessing performance based on cross-validation</a></li>
  <li><a href="#applying-machine-learning-algorithm-with-tuned-hyperparameters" id="toc-applying-machine-learning-algorithm-with-tuned-hyperparameters" class="nav-link" data-scroll-target="#applying-machine-learning-algorithm-with-tuned-hyperparameters"><span class="header-section-number">10.4</span> Applying machine-learning algorithm with tuned hyperparameters</a></li>
  <li><a href="#task-2" id="toc-task-2" class="nav-link" data-scroll-target="#task-2"><span class="header-section-number">10.5</span> Task</a></li>
  </ul></li>
  </ul>
</nav>
</div>
<main class="content quarto-banner-title-block" id="quarto-document-content">





<section id="introduction" class="level1" data-number="1">
<h1 data-number="1"><span class="header-section-number">1</span> Introduction</h1>
<p>In the previous workshop, we introduced <a href="https://www.cog-genomics.org/plink/">PLINK</a> to process genetic data for the purposes of data conversion, calculations, filtering, and running a genetic association study of a binary outcome. In this workshop, you will take a deeper dive into performing genetic association testing of common human traits, meta-analysis of results, and accounting for potential confounding caused by differences in genetic ancestry of individuals in a data set.</p>
</section>
<section id="aims-of-workshop" class="level1" data-number="2">
<h1 data-number="2"><span class="header-section-number">2</span> Aims of workshop</h1>
<ul>
<li>reinforce the concepts of genetic associations in the context of common disease, as covered in this week’s lecture</li>
<li>apply regression-based models for association testing</li>
<li>perform meta-analysis of results across multiple studies</li>
<li>highlight the issue of including individuals of different genetic ancestry in genetic association analysis</li>
<li>perform principal component analysis (PCA) using genetics to separate out individuals based on genetic ancestry</li>
<li>apply clustering methods to group individuals of similar characteristics together and label their ancestry</li>
<li>adjust for presence of population stratification in GWAS</li>
<li>understand the limitations of data commonly used for genome-wide associations</li>
</ul>
</section>
<section id="conda-environment" class="level1" data-number="3">
<h1 data-number="3"><span class="header-section-number">3</span> Conda environment</h1>
<p>Before attempting this workshop, you should activate the <code>hpdm208z</code> conda environment that will provide the R through Jupyter Lab and required library dependencies:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate hpdm208z</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="workshop-directory" class="level1" data-number="4">
<h1 data-number="4"><span class="header-section-number">4</span> Workshop directory</h1>
<p>All relevant files associated with this workshops can be found in the following directory:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb2"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="ex">/home/ubuntu/hpdm208z/workshops/gwas/</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>You should navigate to this directory prior to starting the workshop. Note, a jupyter notebook can be be loaded from this directory to assist you in undertaking this workshop:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb3"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="ex">gwas.ipynb</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="workshop-data-and-files" class="level1" data-number="5">
<h1 data-number="5"><span class="header-section-number">5</span> Workshop data and files</h1>
<p>Below is a table describing the files in the <code>~/hpdm208z/workshops/gwas/</code> directory:</p>
<table class="caption-top table">
<colgroup>
<col style="width: 30%">
<col style="width: 70%">
</colgroup>
<thead>
<tr class="header">
<th>File Name</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>height_study1.[bed|bim|fam]</code></td>
<td>Genetic data on 10,000 SNPs in 10,000 individuals. Height (cm) in .fam file</td>
</tr>
<tr class="even">
<td><code>height_study1.covar</code></td>
<td>Data on sex and age recorded at study enrollment</td>
</tr>
<tr class="odd">
<td><code>height_study2.[bed|bim|fam]</code></td>
<td>Genetic data on 10,000 SNPs in 10,000 individuals. Height (cm) in .fam file</td>
</tr>
<tr class="even">
<td><code>height_study2.covar</code></td>
<td>Data on sex and age recorded at study enrollment</td>
</tr>
<tr class="odd">
<td></td>
<td></td>
</tr>
<tr class="even">
<td><code>METAANALYZE.py</code></td>
<td>Python script for meta-analysis of GWAS statistics from multiple studies</td>
</tr>
<tr class="odd">
<td></td>
<td></td>
</tr>
<tr class="even">
<td><code>ref_samples_for_pca.txt</code></td>
<td>100,000 SNP genotypes for 1,223 individuals &amp; PCA analysis</td>
</tr>
<tr class="odd">
<td><code>study_samples_for_pca.txt</code></td>
<td>100,000 SNP genotypes for 1,222 individuals &amp; PCA analysis</td>
</tr>
<tr class="even">
<td><code>world_cup_fans.[bed|bim|fam]</code></td>
<td>40,938 SNP genotypes with MAF &gt;5% for 1,222 individuals &amp; GWAS</td>
</tr>
</tbody>
</table>
<hr>
</section>
<section id="running-a-gwas-for-adult-human-height" class="level1" data-number="6">
<h1 data-number="6"><span class="header-section-number">6</span> Running a GWAS for adult human height</h1>
<p>We will first use PLINK to perform basic association testing of genetic variants against a continuous phenotype (adult height) while adjusting for covariates that can influence the outcome (age and sex).</p>
<p>We will ask PLINK to perform a linear-regression based analyses for each genetic variant. We will require the following options:</p>
<ul>
<li><code>--bfile</code> - file prefix of the .bed .bim .fam files containing data on genotypes. The phenotype is held in the last column of the .fam file</li>
<li><code>--linear hide-covar</code> - specifies we want to run a linear regression model. <code>hide-covar</code> tells plink not to output the association results for the covariates - only the genetic variant for each model.</li>
<li><code>--covar</code> - name of file containing data on covariates (age and sex) for each individual</li>
<li><code>--out</code> - specifies the primary name of the output file.</li>
</ul>
<p>Run the following command:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb4"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="ex">~/hpdm208z/programs/plink</span> <span class="dt">\</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">--bfile</span> height_study1 <span class="dt">\</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">--linear</span> hide-covar <span class="dt">\</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">--covar</span> height_study1.covar <span class="dt">\</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">--out</span> height_study1_gwas</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>Using the command above, PLINK will generate an output file called <code>height_study1_gwas.assoc.linear</code>. Let’s take a look at it:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb5"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span> height_study1_gwas.assoc.linear</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb6"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a> <span class="ex">CHR</span>         SNP         BP   A1       TEST    NMISS       BETA         STAT            P </span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>   <span class="ex">1</span>   1:1001000    1001000    C        ADD    10000    <span class="at">-0.7364</span>       <span class="at">-1.249</span>       0.2116</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>   <span class="ex">1</span>   1:1003000    1003000    C        ADD    10000      0.882        1.604       0.1088</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>   <span class="ex">1</span>   1:1005000    1005000    G        ADD    10000   <span class="at">-0.01133</span>     <span class="at">-0.02728</span>       0.9782</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>   <span class="ex">1</span>   1:1007000    1007000    G        ADD    10000    <span class="at">-0.1459</span>      <span class="at">-0.7232</span>       0.4696</span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>   <span class="ex">1</span>   1:1009000    1009000    A        ADD    10000     0.4892       0.8272       0.4081</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>   <span class="ex">1</span>   1:1011000    1011000    G        ADD    10000     <span class="at">-0.298</span>       <span class="at">-1.821</span>      0.06858</span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>   <span class="ex">1</span>   1:1013000    1013000    G        ADD    10000   <span class="at">-0.08145</span>      <span class="at">-0.6061</span>       0.5444</span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a>   <span class="ex">1</span>   1:1015000    1015000    T        ADD    10000    0.07354       0.1666       0.8676</span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a>   <span class="ex">1</span>   1:1017000    1017000    A        ADD    10000   <span class="at">-0.05728</span>      <span class="at">-0.3473</span>       0.7284</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>Here is a table that describes the columns in the output</p>
<table class="caption-top table">
<colgroup>
<col style="width: 22%">
<col style="width: 77%">
</colgroup>
<thead>
<tr class="header">
<th>Column</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>CHR</code></td>
<td>From .bim file - chromosome genetic variant resids on</td>
</tr>
<tr class="even">
<td><code>SNP</code></td>
<td>From .bim file - unique name of genetic variant</td>
</tr>
<tr class="odd">
<td><code>BP</code></td>
<td>From .bim file - base-pair position genetic variant found</td>
</tr>
<tr class="even">
<td><code>A1</code></td>
<td>From .bim file - allele assigned A1 (usually minor allele) the BETA corresponds to</td>
</tr>
<tr class="odd">
<td><code>TEST</code></td>
<td>Describes model assumption - <code>ADD</code> = additive model of association</td>
</tr>
<tr class="even">
<td><code>NMISS</code></td>
<td>Number of individuals that went into the analysis as not missing data</td>
</tr>
<tr class="odd">
<td><code>BETA</code></td>
<td>The effect of carrying each additional <code>A1</code> allele on the outcome</td>
</tr>
<tr class="even">
<td><code>STAT</code></td>
<td>Z-statistic calculed as BETA / standard error (not shown) used to obtain P-value</td>
</tr>
<tr class="odd">
<td><code>P</code></td>
<td>P-value indicating strength of statistical significance of estimated effect</td>
</tr>
</tbody>
</table>
<section id="task" class="level2" data-number="6.1">
<h2 data-number="6.1" class="anchored" data-anchor-id="task"><span class="header-section-number">6.1</span> Task</h2>
<ul>
<li>Use PLINK to perform a GWAS analysis using the <code>height_study2.[bed|bim|fam]</code> and <code>height_study1.covar</code> . Output the results to a file called <code>height_study2_gwas.assoc.linear</code>.</li>
</ul>
</section>
</section>
<section id="preparing-results-for-meta-analysis" class="level1" data-number="7">
<h1 data-number="7"><span class="header-section-number">7</span> Preparing results for meta-analysis</h1>
<p>Next, we will attempt to double our effective sample size to 20,000 individuals by meta-analysing these two result sets together using a <strong>fixed-effects inverse-variance weighted</strong> model. A python script has been created for you to perform the meta-analysis of the results from the two studies.</p>
<p>By <em>fixed-effects</em> we just mean that there is only one true effect for a given genetic variant on the phenotypic outcome and differences in effects observed across different studies is due to random sampling of individuals that went into the association analyses.</p>
<p>By <em>inverse-variance weighted</em> we just mean that when we combine studies together, we will give more weighting to larger studies when combining the association statistics for a given genetic variant to calculate an overall statistic. See this week’s notes to remind yourself of how this works.</p>
<section id="ensuring-we-know-which-is-the-effect-allele" class="level2" data-number="7.1">
<h2 data-number="7.1" class="anchored" data-anchor-id="ensuring-we-know-which-is-the-effect-allele"><span class="header-section-number">7.1</span> Ensuring we know which is the effect allele</h2>
<p>However, before we run the Python script you will notice that the output only contains information for 1 of the alleles (A1). The output does not contain the other allele. We need to make sure the alleles are coded the same way across the two studies and potentially account for A1 to be different across the two studies that may result in both an effect with the same magnitude of effect but with different signs of direction of effect (i.e.&nbsp;a positive beta in one study and a negative beta in the other).</p>
<p>Let’s quickly wrangle the data so we are ready to do the meta-analysis.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb7"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Import pandas library</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> sklearn.decomposition <span class="im">import</span> PCA</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>pd.set_option(<span class="st">'display.max_columns'</span>, <span class="va">None</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Read GWAS data for study1 into a pandas df </span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>study1_height_gwas <span class="op">=</span> pd.read_csv(<span class="st">"height_study1_gwas.assoc.linear"</span>, sep<span class="op">=</span><span class="st">"</span><span class="er">\</span><span class="st">s+"</span>)</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Rename "A1" as "EFFECT_ALLELE"</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>study1_height_gwas.rename(columns<span class="op">=</span>{<span class="st">"A1"</span>: <span class="st">"EFFECT_ALLELE"</span>}, inplace<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a><span class="co"># preview</span></span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(study1_height_gwas.head())</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>   CHR        SNP       BP EFFECT_ALLELE TEST  NMISS     BETA     STAT       P
0    1  1:1001000  1001000             C  ADD  10000 -0.73640 -1.24900  0.2116
1    1  1:1003000  1003000             C  ADD  10000  0.88200  1.60400  0.1088
2    1  1:1005000  1005000             G  ADD  10000 -0.01133 -0.02728  0.9782
3    1  1:1007000  1007000             G  ADD  10000 -0.14590 -0.72320  0.4696
4    1  1:1009000  1009000             A  ADD  10000  0.48920  0.82720  0.4081</code></pre>
</div>
</div>
<p>Now we will load in the alleles from the bim file.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load the SNP A1, and A2 columns from the .bim file for study 1 only</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>study1_bim <span class="op">=</span> pd.read_csv(<span class="st">"height_study1.bim"</span>, sep<span class="op">=</span><span class="st">"</span><span class="er">\</span><span class="st">s+"</span>, header<span class="op">=</span><span class="va">None</span>, </span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>                         usecols<span class="op">=</span>[<span class="dv">1</span>, <span class="dv">4</span>, <span class="dv">5</span>], names<span class="op">=</span>[<span class="st">"SNP"</span>, <span class="st">"BIM_A1"</span>, <span class="st">"BIM_A2"</span>])</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a><span class="co"># preview</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(study1_bim.head())</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>         SNP BIM_A1 BIM_A2
0  1:1001000      C      G
1  1:1003000      C      T
2  1:1005000      G      T
3  1:1007000      G      C
4  1:1009000      A      T</code></pre>
</div>
</div>
<p>Merge the GWAS association statistics with the data frame containing the other allele.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb12"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>study1_merged <span class="op">=</span> pd.merge(study1_bim, study1_height_gwas, on<span class="op">=</span><span class="st">"SNP"</span>)</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(study1_merged)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>             SNP BIM_A1 BIM_A2  CHR       BP EFFECT_ALLELE TEST  NMISS  \
0      1:1001000      C      G    1  1001000             C  ADD  10000   
1      1:1003000      C      T    1  1003000             C  ADD  10000   
2      1:1005000      G      T    1  1005000             G  ADD  10000   
3      1:1007000      G      C    1  1007000             G  ADD  10000   
4      1:1009000      A      T    1  1009000             A  ADD  10000   
...          ...    ...    ...  ...      ...           ...  ...    ...   
9995  22:1336000      G      T   22  1336000             G  ADD  10000   
9996  22:1338000      A      G   22  1338000             A  ADD  10000   
9997  22:1340000      C      A   22  1340000             C  ADD  10000   
9998  22:1342000      C      A   22  1342000             C  ADD  10000   
9999  22:1344000      A      C   22  1344000             A  ADD  10000   

         BETA     STAT        P  
0    -0.73640 -1.24900  0.21160  
1     0.88200  1.60400  0.10880  
2    -0.01133 -0.02728  0.97820  
3    -0.14590 -0.72320  0.46960  
4     0.48920  0.82720  0.40810  
...       ...      ...      ...  
9995 -0.03786 -0.19500  0.84540  
9996  0.34380  1.75400  0.07942  
9997 -0.14930 -1.08800  0.27660  
9998 -0.17390 -0.73830  0.46040  
9999  0.12440  0.95590  0.33920  

[10000 rows x 11 columns]</code></pre>
</div>
</div>
<p>Now let’s create a column called <code>OTHER_ALLELE</code> in the GWAS dataset based on <code>EFFECT_ALLELE</code> and two alleles in the .bim file</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb14"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>study1_merged[<span class="st">"OTHER_ALLELE"</span>] <span class="op">=</span> np.where(study1_merged[<span class="st">"EFFECT_ALLELE"</span>] <span class="op">==</span> study1_merged[<span class="st">"BIM_A1"</span>], </span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>                                         study1_merged[<span class="st">"BIM_A2"</span>], study1_merged[<span class="st">"BIM_A1"</span>])</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>                                         </span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(study1_merged)                                         </span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>             SNP BIM_A1 BIM_A2  CHR       BP EFFECT_ALLELE TEST  NMISS  \
0      1:1001000      C      G    1  1001000             C  ADD  10000   
1      1:1003000      C      T    1  1003000             C  ADD  10000   
2      1:1005000      G      T    1  1005000             G  ADD  10000   
3      1:1007000      G      C    1  1007000             G  ADD  10000   
4      1:1009000      A      T    1  1009000             A  ADD  10000   
...          ...    ...    ...  ...      ...           ...  ...    ...   
9995  22:1336000      G      T   22  1336000             G  ADD  10000   
9996  22:1338000      A      G   22  1338000             A  ADD  10000   
9997  22:1340000      C      A   22  1340000             C  ADD  10000   
9998  22:1342000      C      A   22  1342000             C  ADD  10000   
9999  22:1344000      A      C   22  1344000             A  ADD  10000   

         BETA     STAT        P OTHER_ALLELE  
0    -0.73640 -1.24900  0.21160            G  
1     0.88200  1.60400  0.10880            T  
2    -0.01133 -0.02728  0.97820            T  
3    -0.14590 -0.72320  0.46960            C  
4     0.48920  0.82720  0.40810            T  
...       ...      ...      ...          ...  
9995 -0.03786 -0.19500  0.84540            T  
9996  0.34380  1.75400  0.07942            G  
9997 -0.14930 -1.08800  0.27660            A  
9998 -0.17390 -0.73830  0.46040            A  
9999  0.12440  0.95590  0.33920            C  

[10000 rows x 12 columns]</code></pre>
</div>
</div>
<p>Now we have both alleles in the data frame defined as <code>EFFECT_ALLELE</code> and <code>OTHER_ALLELE</code>, lets remove the columns we no longer need: <code>BIM_A1</code> and <code>BIM_A2</code>, and the move <code>OTHER_ALLELE</code> column next to the <code>EFFECT_ALLELE</code> column in the data frame:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb16"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>study1_merged <span class="op">=</span> study1_merged.drop(columns<span class="op">=</span>[<span class="st">"BIM_A1"</span>, <span class="st">"BIM_A2"</span>])</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>col <span class="op">=</span> study1_merged.pop(<span class="st">"OTHER_ALLELE"</span>)</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>study1_merged.insert(<span class="dv">4</span>, <span class="st">"OTHER_ALLELE"</span>, col)</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a><span class="co"># preview</span></span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(study1_merged)  </span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>             SNP  CHR       BP EFFECT_ALLELE OTHER_ALLELE TEST  NMISS  \
0      1:1001000    1  1001000             C            G  ADD  10000   
1      1:1003000    1  1003000             C            T  ADD  10000   
2      1:1005000    1  1005000             G            T  ADD  10000   
3      1:1007000    1  1007000             G            C  ADD  10000   
4      1:1009000    1  1009000             A            T  ADD  10000   
...          ...  ...      ...           ...          ...  ...    ...   
9995  22:1336000   22  1336000             G            T  ADD  10000   
9996  22:1338000   22  1338000             A            G  ADD  10000   
9997  22:1340000   22  1340000             C            A  ADD  10000   
9998  22:1342000   22  1342000             C            A  ADD  10000   
9999  22:1344000   22  1344000             A            C  ADD  10000   

         BETA     STAT        P  
0    -0.73640 -1.24900  0.21160  
1     0.88200  1.60400  0.10880  
2    -0.01133 -0.02728  0.97820  
3    -0.14590 -0.72320  0.46960  
4     0.48920  0.82720  0.40810  
...       ...      ...      ...  
9995 -0.03786 -0.19500  0.84540  
9996  0.34380  1.75400  0.07942  
9997 -0.14930 -1.08800  0.27660  
9998 -0.17390 -0.73830  0.46040  
9999  0.12440  0.95590  0.33920  

[10000 rows x 10 columns]</code></pre>
</div>
</div>
</section>
<section id="obtaining-the-standard-error-of-the-effect-estimate" class="level2" data-number="7.2">
<h2 data-number="7.2" class="anchored" data-anchor-id="obtaining-the-standard-error-of-the-effect-estimate"><span class="header-section-number">7.2</span> Obtaining the standard error of the effect estimate</h2>
<p>As we are going to perform a inverse-variance based meta-analysis, we need to obtain the standard error:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb18"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>study1_merged[<span class="st">"SE"</span>] <span class="op">=</span> <span class="bu">round</span>(study1_merged[<span class="st">"BETA"</span>] <span class="op">/</span> study1_merged[<span class="st">"STAT"</span>], <span class="dv">5</span>)</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Reorder columns so that SE is next to the BETA column</span></span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>col <span class="op">=</span> study1_merged.pop(<span class="st">"SE"</span>)</span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>study1_merged.insert(<span class="dv">8</span>, <span class="st">"SE"</span>, col)</span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a><span class="co"># preview</span></span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(study1_merged)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>             SNP  CHR       BP EFFECT_ALLELE OTHER_ALLELE TEST  NMISS  \
0      1:1001000    1  1001000             C            G  ADD  10000   
1      1:1003000    1  1003000             C            T  ADD  10000   
2      1:1005000    1  1005000             G            T  ADD  10000   
3      1:1007000    1  1007000             G            C  ADD  10000   
4      1:1009000    1  1009000             A            T  ADD  10000   
...          ...  ...      ...           ...          ...  ...    ...   
9995  22:1336000   22  1336000             G            T  ADD  10000   
9996  22:1338000   22  1338000             A            G  ADD  10000   
9997  22:1340000   22  1340000             C            A  ADD  10000   
9998  22:1342000   22  1342000             C            A  ADD  10000   
9999  22:1344000   22  1344000             A            C  ADD  10000   

         BETA       SE     STAT        P  
0    -0.73640  0.58959 -1.24900  0.21160  
1     0.88200  0.54988  1.60400  0.10880  
2    -0.01133  0.41532 -0.02728  0.97820  
3    -0.14590  0.20174 -0.72320  0.46960  
4     0.48920  0.59139  0.82720  0.40810  
...       ...      ...      ...      ...  
9995 -0.03786  0.19415 -0.19500  0.84540  
9996  0.34380  0.19601  1.75400  0.07942  
9997 -0.14930  0.13722 -1.08800  0.27660  
9998 -0.17390  0.23554 -0.73830  0.46040  
9999  0.12440  0.13014  0.95590  0.33920  

[10000 rows x 11 columns]</code></pre>
</div>
</div>
<p>Now we can export the data for study1 that can be used for meta-analysis:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb20"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>study1_merged.to_csv(<span class="st">"height_study1_gwas.assoc.linear.formatted"</span>, sep<span class="op">=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, index<span class="op">=</span><span class="va">False</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>Now we need to do the same for the other GWAS summary file from study2</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb21"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Read GWAS data for study1 into a pandas df </span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>study2_height_gwas <span class="op">=</span> pd.read_csv(<span class="st">"height_study2_gwas.assoc.linear"</span>, sep<span class="op">=</span><span class="vs">r"</span><span class="dv">\s</span><span class="op">+</span><span class="vs">"</span>)</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Rename "A1" as "EFFECT_ALLELE"</span></span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>study2_height_gwas.rename(columns<span class="op">=</span>{<span class="st">"A1"</span>: <span class="st">"EFFECT_ALLELE"</span>}, inplace<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Load the SNP and A2 columns from the .bim file for study 1 only</span></span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a>study2_bim <span class="op">=</span> pd.read_csv(<span class="st">"height_study2.bim"</span>, sep<span class="op">=</span><span class="st">"</span><span class="er">\</span><span class="st">s+"</span>, header<span class="op">=</span><span class="va">None</span>, </span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a>                         usecols<span class="op">=</span>[<span class="dv">1</span>, <span class="dv">4</span>, <span class="dv">5</span>], names<span class="op">=</span>[<span class="st">"SNP"</span>, <span class="st">"BIM_A1"</span>, <span class="st">"BIM_A2"</span>])</span>
<span id="cb21-10"><a href="#cb21-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-11"><a href="#cb21-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Merge</span></span>
<span id="cb21-12"><a href="#cb21-12" aria-hidden="true" tabindex="-1"></a>study2_merged <span class="op">=</span> pd.merge(study2_bim, study2_height_gwas, on<span class="op">=</span><span class="st">"SNP"</span>)</span>
<span id="cb21-13"><a href="#cb21-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-14"><a href="#cb21-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Create OTHER_ALLELE column</span></span>
<span id="cb21-15"><a href="#cb21-15" aria-hidden="true" tabindex="-1"></a>study2_merged[<span class="st">"OTHER_ALLELE"</span>] <span class="op">=</span> np.where(study2_merged[<span class="st">"EFFECT_ALLELE"</span>] <span class="op">==</span> study2_merged[<span class="st">"BIM_A1"</span>], </span>
<span id="cb21-16"><a href="#cb21-16" aria-hidden="true" tabindex="-1"></a>                                         study2_merged[<span class="st">"BIM_A2"</span>], study2_merged[<span class="st">"BIM_A1"</span>])</span>
<span id="cb21-17"><a href="#cb21-17" aria-hidden="true" tabindex="-1"></a>                                   </span>
<span id="cb21-18"><a href="#cb21-18" aria-hidden="true" tabindex="-1"></a><span class="co"># Drop helper columns </span></span>
<span id="cb21-19"><a href="#cb21-19" aria-hidden="true" tabindex="-1"></a>study2_merged <span class="op">=</span> study2_merged.drop(columns<span class="op">=</span>[<span class="st">"BIM_A1"</span>, <span class="st">"BIM_A2"</span>])</span>
<span id="cb21-20"><a href="#cb21-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-21"><a href="#cb21-21" aria-hidden="true" tabindex="-1"></a><span class="co"># Move OTHER_ALLELE next to EFFECT_ALLELE in data frame</span></span>
<span id="cb21-22"><a href="#cb21-22" aria-hidden="true" tabindex="-1"></a>col <span class="op">=</span> study2_merged.pop(<span class="st">"OTHER_ALLELE"</span>)</span>
<span id="cb21-23"><a href="#cb21-23" aria-hidden="true" tabindex="-1"></a>study2_merged.insert(<span class="dv">4</span>, <span class="st">"OTHER_ALLELE"</span>, col)</span>
<span id="cb21-24"><a href="#cb21-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-25"><a href="#cb21-25" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate the SE from the BETA and STAT values</span></span>
<span id="cb21-26"><a href="#cb21-26" aria-hidden="true" tabindex="-1"></a>study2_merged[<span class="st">"SE"</span>] <span class="op">=</span> <span class="bu">round</span>(study2_merged[<span class="st">"BETA"</span>] <span class="op">/</span> study2_merged[<span class="st">"STAT"</span>], <span class="dv">5</span>)</span>
<span id="cb21-27"><a href="#cb21-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-28"><a href="#cb21-28" aria-hidden="true" tabindex="-1"></a><span class="co"># Reorder columns so that SE is next to the BETA column</span></span>
<span id="cb21-29"><a href="#cb21-29" aria-hidden="true" tabindex="-1"></a>col <span class="op">=</span> study2_merged.pop(<span class="st">"SE"</span>)</span>
<span id="cb21-30"><a href="#cb21-30" aria-hidden="true" tabindex="-1"></a>study2_merged.insert(<span class="dv">8</span>, <span class="st">"SE"</span>, col)</span>
<span id="cb21-31"><a href="#cb21-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-32"><a href="#cb21-32" aria-hidden="true" tabindex="-1"></a><span class="co"># EXPORT</span></span>
<span id="cb21-33"><a href="#cb21-33" aria-hidden="true" tabindex="-1"></a>study2_merged.to_csv(<span class="st">"height_study2_gwas.assoc.linear.formatted"</span>, sep<span class="op">=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, index<span class="op">=</span><span class="va">False</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
</section>
<section id="meta-analysis-of-the-gwas-data" class="level1" data-number="8">
<h1 data-number="8"><span class="header-section-number">8</span> Meta-analysis of the GWAS data</h1>
<p>Now the two results have both alleles and the standard error, we can now proceed with the meta-analysis.</p>
<p>A Python script has been provided to enable you to meta-analyse genetic association statistics across studies</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb22"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="ex">python</span> METAANALYZE.py <span class="dt">\</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">--files</span> height_study1_gwas.assoc.linear.formatted,height_study2_gwas.assoc.linear.formatted <span class="dt">\</span></span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">--markername</span> SNP <span class="dt">\</span></span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">--effect-allele</span> EFFECT_ALLELE <span class="dt">\</span></span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">--other-allele</span> OTHER_ALLELE <span class="dt">\</span></span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">--effect</span> BETA <span class="dt">\</span></span>
<span id="cb22-7"><a href="#cb22-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">--se</span> SE <span class="dt">\</span></span>
<span id="cb22-8"><a href="#cb22-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">--out</span> HEIGHT_MA.txt</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<section id="question" class="level2" data-number="8.1">
<h2 data-number="8.1" class="anchored" data-anchor-id="question"><span class="header-section-number">8.1</span> Question</h2>
<ul>
<li>How does the number of genome-wide significant associations with P&lt;5x10-8 compare when looking at the results from each study (<code>height_study1_gwas.assoc.linear.formatted</code> and <code>height_study2_gwas.assoc.linear.formatted</code>) compared to the meta-analysis results held in <code>HEIGHT_MA.txt</code>?</li>
</ul>
</section>
</section>
<section id="dealing-with-population-stratification" class="level1" data-number="9">
<h1 data-number="9"><span class="header-section-number">9</span> Dealing with population stratification</h1>
<p>The allele frequencies of many genetic variants are not constant across different parts of the world - some alleles are more frequent among some populations compared to others. If there are differences in <strong>genetic ancestry</strong> among individuals included in a single GWAS analysis, then there is potential for false-positive associations to arise.</p>
<p>In the context of a disease, this can happen when there is a both an over-representation of disease cases from a genetic ancestry and systematic differences in allele frequencies between that ancestry and other ancestries. This introduces spurious correlations between genotypes and disease status. For continuous traits, false-positive associations could occur when higher levels of the trait are over-represented by a genetic ancestry with higher allele frequencies compared to other genetic ancestries.</p>
<p>These scenarios are referred to as <strong>population stratification</strong> and should be accounted for when running analyses.</p>
<p>In the following analyses we will use genetics to estimate the genetic ancestral similarity of individuals and relative differences among individuals using an unsupervised machine learning method - <strong>principal components analysis (PCA)</strong>.</p>
<section id="football-world-cup" class="level2" data-number="9.1">
<h2 data-number="9.1" class="anchored" data-anchor-id="football-world-cup"><span class="header-section-number">9.1</span> Football World Cup</h2>
<p>To demonstrate the requirement to handle population stratification, we are going to perform a GWAS based on data from a questionnaire that asked individuals who went to the FIFA World Cup 2022 in Qatar whether they preferred to watch a match involving a team from East Asia or from somewhere else.</p>
<p>In our scenario, the estimated heritability for match preference is zero i.e.&nbsp;there is no meaningful biology altered through genetic variation that contributed to the variability in whether someone preferred to travel to watch a team from East Asia play. Therefore, we would not expect to identify genetic variants through GWAS associated with such preference.</p>
<p>If an individual reported that they preferred attending a match fixture involving a team from East Asia, they were assigned the value <code>1</code>, otherwise <code>0</code>. Note, you may assume that all participants who answered the questionnaire had the same opportunity to attend any match at the World Cup.</p>
</section>
<section id="task---gwas-of-east-asian-football-team-support" class="level2" data-number="9.2">
<h2 data-number="9.2" class="anchored" data-anchor-id="task---gwas-of-east-asian-football-team-support"><span class="header-section-number">9.2</span> Task - GWAS of East-Asian football team support</h2>
<p>Perform a GWAS of East-Asia vs rest of the world match preference using following command:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb23"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="ex">~/hpdm208z/programs/plink</span> <span class="dt">\</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">--bfile</span> world_cup_fans <span class="dt">\</span></span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">--1</span> <span class="dt">\</span></span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">--logistic</span> hide-covar <span class="dt">\</span></span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">--out</span> world_cup_east_asia_team_preference</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>Let’s take a look at the number of genetic variants reaching genome-wide level signficance based on a P-value based on 5x10-8.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb24"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>world_cup_gwas_results <span class="op">=</span> pd.read_csv(<span class="st">"world_cup_east_asia_team_preference.assoc.logistic"</span>, sep<span class="op">=</span><span class="vs">r"</span><span class="dv">\s</span><span class="op">+</span><span class="vs">"</span>)</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a><span class="co"># preview</span></span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(world_cup_gwas_results.head())</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>   CHR       SNP      BP A1 TEST  NMISS      OR   STAT         P
0    1   1:14599   14599  A  ADD   1222  0.6090 -2.902  0.003703
1    1   1:14930   14930  G  ADD   1222  0.4892 -3.526  0.000423
2    1   1:15820   15820  T  ADD   1222  1.4780  3.735  0.000188
3    1   1:86331   86331  G  ADD   1222  1.2920  1.698  0.089540
4    1  1:104186  104186  T  ADD   1222  1.6790  3.757  0.000172</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb26"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Subset to genome-wide significant and order by P-value</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>world_cup_gwas_results_sig <span class="op">=</span> world_cup_gwas_results.loc[world_cup_gwas_results[<span class="st">'P'</span>] <span class="op">&lt;</span> <span class="fl">5e-08</span>]</span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Order by P and print top 10 results</span></span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(world_cup_gwas_results_sig.sort_values(<span class="st">"P"</span>).head())</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>       CHR          SNP         BP A1 TEST  NMISS      OR   STAT             P
4353     2   2:55577253   55577253  T  ADD   1222  2.7440  9.018  1.911000e-19
5378     2  2:145753166  145753166  T  ADD   1222  2.5380  8.856  8.260000e-19
12084    5   5:29436064   29436064  C  ADD   1222  2.6290  8.716  2.893000e-18
18874    8   8:21430468   21430468  C  ADD   1222  5.2570  8.602  7.822000e-18
15664    6  6:151265412  151265412  A  ADD   1222  0.3819 -8.589  8.792000e-18</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb28"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Count number of GWAS significant</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="bu">len</span>(world_cup_gwas_results_sig))</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>1619</code></pre>
</div>
</div>
<p>A total of <strong>1,619</strong> SNPs reached genome-wide significance for a trait that has 0% heritability!</p>
<p>We might hypothesize that some of most significant variants have alleles that are more common among individuals of East Asian (EAS) genetic ancestry over others. For example, one of the most significant genetic variants is <code>2:145753166</code> This variant is described <a href="http://grch37.ensembl.org/Homo_sapiens/Variation/Population?db=core;r=2:145753166-145753166;v=rs1371048;vdb=variation;vf=144460854">here</a>. The rsid of this variant is <code>rs1371048</code>. The alleles of this SNP in our data are T and G. Based on 1000 Genomes Project data, the T allele is more common in EAS than any of the other populations (and so the G allele most be less common in EAS compared to other populations). Here are the allele frequencies across the 5 main ancestral populations:</p>
<table class="caption-top table">
<thead>
<tr class="header">
<th style="text-align: center;">Allele</th>
<th style="text-align: center;">AFR</th>
<th style="text-align: center;">AMR</th>
<th style="text-align: center;">EAS</th>
<th style="text-align: center;">EUR</th>
<th style="text-align: center;">SAS</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">T</td>
<td style="text-align: center;">21%</td>
<td style="text-align: center;">56%</td>
<td style="text-align: center;">98%</td>
<td style="text-align: center;">34%</td>
<td style="text-align: center;">50%</td>
</tr>
<tr class="even">
<td style="text-align: center;">G</td>
<td style="text-align: center;">79%</td>
<td style="text-align: center;">44%</td>
<td style="text-align: center;">2%</td>
<td style="text-align: center;">66%</td>
<td style="text-align: center;">50%</td>
</tr>
</tbody>
</table>
</section>
<section id="labelling-and-quantifying-differences-in-genetic-ancestry" class="level2" data-number="9.3">
<h2 data-number="9.3" class="anchored" data-anchor-id="labelling-and-quantifying-differences-in-genetic-ancestry"><span class="header-section-number">9.3</span> Labelling and quantifying differences in genetic ancestry</h2>
<p>We are going to perform principal components analysis (PCA) to separate individuals out quantitatively based on genetic differences and similarities. We will perform the analysis in two steps:</p>
<ol type="1">
<li>Create a principal components space using a reference data set (the 1000 Genomes Project) with labelled individuals from different genetic ancestries.</li>
<li>Project individuals from our study into the PC-space.</li>
</ol>
<p>Once we have a numerical representation of ancestral similarities, we will adjust the GWAS for the principal component values to account for differences in genetic ancestry in our genetic association analysis. Later, we will additionally label individuals as belonging to a broad ancestry group using a supervised machine learning algorithm.</p>
</section>
<section id="performing-pca-on-the-reference-dataset" class="level2" data-number="9.4">
<h2 data-number="9.4" class="anchored" data-anchor-id="performing-pca-on-the-reference-dataset"><span class="header-section-number">9.4</span> Performing PCA on the reference dataset</h2>
<p>The first task is to create a PC-space using reference data obtained from the 1000 Genomes project.</p>
<p>We are going to use Python to perform the PCA using 100,000 SNP genotypes for 1,223 individuals known to be from a mixture of broad genetic ancestries:</p>
<ul>
<li><strong>AFR</strong> : African</li>
<li><strong>AMR</strong> : Ad-mixed American</li>
<li><strong>EAS</strong> : East-Asian</li>
<li><strong>EUR</strong> : European</li>
<li><strong>SAS</strong> : South-Asian</li>
</ul>
<p>First, create a Pandas DF called <code>ref_samples</code> that contains the population code and sample ID in the columns <code>FID</code> and <code>IID</code>, respectively. We will merge the calculated principal component values associated with each sample with this DF later on.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb30"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>ref_samples <span class="op">=</span> pd.read_csv(<span class="st">"ref_samples_for_pca.txt"</span>, usecols<span class="op">=</span>[<span class="st">"FID"</span>,<span class="st">"IID"</span>], delimiter<span class="op">=</span><span class="st">" "</span>)</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>ref_samples.head()</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>   FID      IID
0  EUR  HG00097
1  EUR  HG00099
2  EUR  HG00101
3  EUR  HG00102
4  EUR  HG00105</code></pre>
</div>
</div>
<p>Now load the genotypes of 100,000 SNPs in 1,223 individuals found in the reference data into 2D numpy array.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb32"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Skip header, and extract columns 3 - 100,002 (indexed as 2 to 100001)</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>X1 <span class="op">=</span> np.loadtxt(fname<span class="op">=</span><span class="st">"ref_samples_for_pca.txt"</span>, skiprows<span class="op">=</span><span class="dv">1</span>, usecols<span class="op">=</span>np.arange(<span class="dv">2</span>,<span class="dv">100002</span>)) </span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>X1.shape </span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>(1223, 100000)</code></pre>
</div>
</div>
<p>We will now perform PCA using the reference dataset and return the first 10 PCs using the library:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb34"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>pca <span class="op">=</span> PCA(n_components<span class="op">=</span><span class="dv">10</span>)</span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>Y1 <span class="op">=</span> pca.fit_transform(X1)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p><code>Y1</code> is a 2D numpy array containing the values of the first 10 principal components for each individual.</p>
<p>Note that is common to scale and mean-centre data prior to performing PCA analysis. However, we do not need to provide explicit code for this because:</p>
<ul>
<li>all 100,000 variables are on the same scale so do not need to be re-scaled</li>
<li>the <code>PCA()</code> function implemented in the <code>scikit-learn</code> library carries out mean-centering of the data automatically, prior to the generation of PCs.</li>
</ul>
<p>Let’s merge the <code>ref_samples</code> DF and the numpy array <code>Y1</code> (which we will convert to a Pandas DF) together to visualise the data. Note rows in the matrix containing the PC values are in the same order as the rows in the genotype reference data, making it easy to merge IDs:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb35"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>ref_sample_pcs <span class="op">=</span> pd.concat([ref_samples, pd.DataFrame(Y1)], axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>ref_sample_pcs.columns <span class="op">=</span> [<span class="st">"FID"</span>,<span class="st">"IID"</span>,<span class="st">"PC1"</span>,<span class="st">"PC2"</span>,<span class="st">"PC3"</span>,<span class="st">"PC4"</span>,<span class="st">"PC5"</span>,<span class="st">"PC6"</span>,<span class="st">"PC7"</span>,<span class="st">"PC8"</span>,<span class="st">"PC9"</span>,<span class="st">"PC10"</span>]</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>ref_sample_pcs</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>      FID      IID        PC1        PC2        PC3        PC4        PC5  \
0     EUR  HG00097 -14.426319  33.251409  -9.001286   9.522424  -9.248659   
1     EUR  HG00099 -13.755650  34.191333  -9.868827  10.543570 -10.380949   
2     EUR  HG00101 -15.388023  32.814107 -10.876452  11.691985  -6.806517   
3     EUR  HG00102 -14.205595  34.352392  -9.414806  10.879079  -5.199596   
4     EUR  HG00105 -13.501380  34.263998 -10.330081  11.573443  -8.856633   
...   ...      ...        ...        ...        ...        ...        ...   
1218  SAS  NA21126 -14.292971  11.482773  25.966832  -7.466801   6.907043   
1219  SAS  NA21129 -15.448805  11.583855  26.828434  -6.437027   8.954984   
1220  SAS  NA21130 -16.056254  12.398727  24.178645  -4.723237   9.322577   
1221  SAS  NA21137 -14.480684  12.466331  24.084101  -7.164242   8.024390   
1222  SAS  NA21144 -15.443296  10.985882  24.076536  -3.940127  17.912887   

           PC6        PC7       PC8       PC9       PC10  
0    -0.878138  -4.520736 -3.365676 -4.221558   1.107865  
1    -1.693816  -5.966099 -3.783902 -3.527381   0.614973  
2    -0.587681  -2.750775 -5.732242 -3.818476   2.210309  
3    -1.299292  -3.553589 -5.761407 -3.533703   0.833023  
4    -0.963731  -2.920343 -3.335881 -2.428733  -1.363890  
...        ...        ...       ...       ...        ...  
1218 -1.765146  -9.179737  0.308128 -4.197811 -16.341620  
1219 -2.750567 -10.999121  2.704058 -2.335698 -17.395517  
1220  1.303396  -7.328739 -3.733178  6.828819 -15.537615  
1221 -3.750932  -9.624530  3.078379 -6.341004 -17.904236  
1222  1.315217  -7.869899  1.247662  4.305544 -17.525172  

[1223 rows x 12 columns]</code></pre>
</div>
</div>
<p>Now we can plot the PC values of each individual, colour coded by the genetic ancestry of the individual as recorded by the 1000 Genomes project. Let’s start by plotting PC1 against PC2.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb37"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>ref_groups <span class="op">=</span> ref_sample_pcs.groupby(<span class="st">"FID"</span>)</span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>fig, ax <span class="op">=</span> plt.subplots(figsize <span class="op">=</span> (<span class="dv">10</span>, <span class="dv">8</span>))</span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>ax.margins(<span class="fl">0.05</span>)</span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> name, group <span class="kw">in</span> ref_groups:</span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a>    ax.plot(group.PC1, group.PC2, marker<span class="op">=</span><span class="st">"x"</span>, linestyle<span class="op">=</span><span class="st">""</span>, ms<span class="op">=</span><span class="dv">5</span>, label<span class="op">=</span>name)</span>
<span id="cb37-6"><a href="#cb37-6" aria-hidden="true" tabindex="-1"></a>ax.legend()</span>
<span id="cb37-7"><a href="#cb37-7" aria-hidden="true" tabindex="-1"></a>plt.title(<span class="st">"Principal Component Space for Individuals in 1000G Project: PC1 vs PC2"</span>)</span>
<span id="cb37-8"><a href="#cb37-8" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">"PC1"</span>)</span>
<span id="cb37-9"><a href="#cb37-9" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">"PC2"</span>)</span>
<span id="cb37-10"><a href="#cb37-10" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="hpdm208z_workshop8_files/figure-html/unnamed-chunk-24-1.png" class="img-fluid figure-img" width="960"></p>
</figure>
</div>
</div>
</div>
<section id="task-1" class="level3" data-number="9.4.1">
<h3 data-number="9.4.1" class="anchored" data-anchor-id="task-1"><span class="header-section-number">9.4.1</span> Task</h3>
<ul>
<li>Produce two additional plots: PC3 against PC4, and PC5 against PC6.</li>
</ul>
</section>
</section>
<section id="projecting-study-samples-into-the-reference-pc-space" class="level2" data-number="9.5">
<h2 data-number="9.5" class="anchored" data-anchor-id="projecting-study-samples-into-the-reference-pc-space"><span class="header-section-number">9.5</span> Projecting study samples into the reference PC space</h2>
<p>Now that we have defined a principal component space for the reference data we have from the 1000 Genomes Project, we will now project new samples into the same space. However, <strong>we must ensure the following</strong>:</p>
<ul>
<li>the same 100,000 variables (i.e.&nbsp;SNPs) found in the reference data are available in the new sample data and in the same column order</li>
<li>the numerical representation of genotypes for SNPs in the new sample data is the <strong>same</strong> as the reference data. For example, an A/G SNP coded as <code>0=AA, 1=AG, 2=GG</code> in the reference data must be coded the same way in the new sample data (so not <code>0=GG, 1=AG, 2=AA</code>). Note, the alleles are already aligned in this workshop so we don’t have to worry about this here.</li>
</ul>
<p>First, load in the study genotypes into Pandas DFs as before - one to hold genetic ancestry (currently unknown) and the unique identifier of each individual. Note the file to be read in is of the same format as the reference dataset.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb38"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>study_samples <span class="op">=</span> pd.read_csv(<span class="st">"study_samples_for_pca.txt"</span>, usecols<span class="op">=</span>[<span class="st">"FID"</span>, <span class="st">"IID"</span>], delimiter<span class="op">=</span><span class="st">" "</span>)</span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>study_samples.head()</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>       FID      IID
0  HG00096  HG00096
1  HG00100  HG00100
2  HG00103  HG00103
3  HG00109  HG00109
4  HG00112  HG00112</code></pre>
</div>
</div>
<p>As before, we need create a 2D numpy array to store the genoypes:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb40"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>X2 <span class="op">=</span> np.loadtxt(fname<span class="op">=</span><span class="st">"study_samples_for_pca.txt"</span>, skiprows<span class="op">=</span><span class="dv">1</span>, usecols<span class="op">=</span>np.arange(<span class="dv">2</span>,<span class="dv">100002</span>))</span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>X2.shape</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>(1222, 100000)</code></pre>
</div>
</div>
<p>We are now able to project the new samples into the principal components space and obtain principal component values for each individual:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb42"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create 2D numpy array containing PC values for study samples by fitting PCA model above.</span></span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>Y2 <span class="op">=</span> pca.transform(X2)</span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-4"><a href="#cb42-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract PCs from Y2 and link up to study IDs</span></span>
<span id="cb42-5"><a href="#cb42-5" aria-hidden="true" tabindex="-1"></a>study_sample_pcs <span class="op">=</span> pd.concat([study_samples, pd.DataFrame(Y2)], axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb42-6"><a href="#cb42-6" aria-hidden="true" tabindex="-1"></a>study_sample_pcs.columns <span class="op">=</span> [<span class="st">"FID"</span>,<span class="st">"IID"</span>,<span class="st">"PC1"</span>,<span class="st">"PC2"</span>,<span class="st">"PC3"</span>,<span class="st">"PC4"</span>,<span class="st">"PC5"</span>,<span class="st">"PC6"</span>,<span class="st">"PC7"</span>,<span class="st">"PC8"</span>,<span class="st">"PC9"</span>,<span class="st">"PC10"</span>]</span>
<span id="cb42-7"><a href="#cb42-7" aria-hidden="true" tabindex="-1"></a>study_sample_pcs</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>          FID      IID        PC1        PC2        PC3        PC4        PC5  \
0     HG00096  HG00096 -13.542557  32.377174 -10.895992  10.696762  -3.575763   
1     HG00100  HG00100 -14.004313  32.241526 -11.099382  13.149988  -1.568393   
2     HG00103  HG00103 -14.483089  32.607163  -9.351953  10.536957  -3.959848   
3     HG00109  HG00109 -15.951840  32.533731 -10.473870  12.378976   0.761252   
4     HG00112  HG00112 -15.633836  33.225539  -8.937601  11.162656  -3.193370   
...       ...      ...        ...        ...        ...        ...        ...   
1217  NA21133  NA21133 -14.312609  10.960837  22.269198  -5.879760   6.244087   
1218  NA21135  NA21135 -14.455674  10.530066  23.284110  -5.778282   6.265334   
1219  NA21141  NA21141 -15.601229  10.736895  22.619389  -3.750600  14.203425   
1220  NA21142  NA21142 -15.895942  10.651730  21.200307  -2.984751  13.984097   
1221  NA21143  NA21143 -16.582568   9.514909  22.259031  -4.312041  12.234487   

           PC6       PC7       PC8       PC9      PC10  
0    -1.749357 -1.758583 -0.597929 -2.754827  3.484493  
1    -0.865930  1.192822  0.566558 -2.792040 -3.372379  
2    -1.071685 -2.878872 -1.471800 -3.045611  2.145038  
3     3.072384  1.575493 -2.918733  3.214377 -1.876545  
4    -0.300740 -1.794649 -3.456817 -0.846815  0.915893  
...        ...       ...       ...       ...       ...  
1217  0.755760 -4.808882 -2.705048  2.953586 -7.807425  
1218  1.198882 -4.817298 -2.986732  3.898059 -7.458517  
1219  1.418397 -3.416132  0.227276  1.304175 -7.576578  
1220  3.393319 -3.191071 -0.544244  0.985557 -8.488138  
1221  1.617300 -3.690932 -0.241286  2.365615 -7.326827  

[1222 rows x 12 columns]</code></pre>
</div>
</div>
<p>We can now plot the reference data with the new samples overlaid. For example, to plot PC1 vs PC2 again, we can use the following code:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb44"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>fig, ax <span class="op">=</span> plt.subplots(figsize <span class="op">=</span> (<span class="dv">10</span>, <span class="dv">8</span>))</span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>ax.margins(<span class="fl">0.05</span>)</span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> name, group <span class="kw">in</span> ref_groups:</span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a>    ax.plot(group.PC1, group.PC2, marker<span class="op">=</span><span class="st">"x"</span>, linestyle<span class="op">=</span><span class="st">""</span>, ms<span class="op">=</span><span class="dv">5</span>, label<span class="op">=</span>name)</span>
<span id="cb44-5"><a href="#cb44-5" aria-hidden="true" tabindex="-1"></a>ax.plot(study_sample_pcs[<span class="st">"PC1"</span>], study_sample_pcs[<span class="st">"PC2"</span>], marker<span class="op">=</span><span class="st">"o"</span>, c<span class="op">=</span><span class="st">"black"</span>, </span>
<span id="cb44-6"><a href="#cb44-6" aria-hidden="true" tabindex="-1"></a>            linestyle<span class="op">=</span><span class="st">""</span>, alpha<span class="op">=</span><span class="fl">0.15</span>, ms<span class="op">=</span><span class="dv">5</span>, label<span class="op">=</span><span class="st">"Study samples"</span>)</span>
<span id="cb44-7"><a href="#cb44-7" aria-hidden="true" tabindex="-1"></a>ax.legend()</span>
<span id="cb44-8"><a href="#cb44-8" aria-hidden="true" tabindex="-1"></a>plt.title(<span class="st">"Principal Component Space for Individuals in 1000G Project and Study: PC1 vs PC2"</span>)</span>
<span id="cb44-9"><a href="#cb44-9" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">"PC1"</span>)</span>
<span id="cb44-10"><a href="#cb44-10" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">"PC2"</span>)</span>
<span id="cb44-11"><a href="#cb44-11" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="hpdm208z_workshop8_files/figure-html/unnamed-chunk-28-3.png" class="img-fluid figure-img" width="960"></p>
</figure>
</div>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb45"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>fig, ax <span class="op">=</span> plt.subplots(figsize <span class="op">=</span> (<span class="dv">10</span>, <span class="dv">8</span>))</span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>ax.margins(<span class="fl">0.05</span>)</span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> name, group <span class="kw">in</span> ref_groups:</span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a>    ax.plot(group.PC3, group.PC4, marker<span class="op">=</span><span class="st">"x"</span>, linestyle<span class="op">=</span><span class="st">""</span>, ms<span class="op">=</span><span class="dv">5</span>, label<span class="op">=</span>name)</span>
<span id="cb45-5"><a href="#cb45-5" aria-hidden="true" tabindex="-1"></a>ax.plot(study_sample_pcs[<span class="st">"PC3"</span>], study_sample_pcs[<span class="st">"PC4"</span>], marker<span class="op">=</span><span class="st">"o"</span>, c<span class="op">=</span><span class="st">"black"</span>, </span>
<span id="cb45-6"><a href="#cb45-6" aria-hidden="true" tabindex="-1"></a>            linestyle<span class="op">=</span><span class="st">""</span>, alpha<span class="op">=</span><span class="fl">0.15</span>, ms<span class="op">=</span><span class="dv">5</span>, label<span class="op">=</span><span class="st">"Study samples"</span>)</span>
<span id="cb45-7"><a href="#cb45-7" aria-hidden="true" tabindex="-1"></a>ax.legend()</span>
<span id="cb45-8"><a href="#cb45-8" aria-hidden="true" tabindex="-1"></a>plt.title(<span class="st">"Principal Component Space for Individuals in 1000G Project and Study: PC3 vs PC4"</span>)</span>
<span id="cb45-9"><a href="#cb45-9" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">"PC3"</span>)</span>
<span id="cb45-10"><a href="#cb45-10" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">"PC4"</span>)</span>
<span id="cb45-11"><a href="#cb45-11" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="hpdm208z_workshop8_files/figure-html/unnamed-chunk-28-4.png" class="img-fluid figure-img" width="960"></p>
</figure>
</div>
</div>
</div>
<p>Now we are ready to extract the PC values for our study samples so that we can subsequently incorporate them as covariates in our GWAS of World Cup ticket purchase preference for East Asia team fixtures. Note, for this analysis we will adjust for the first 5 PCs as they spread out the ancestral populations the most - although we could add more depending on how much variance in the data we want the PC to explain.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb46"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>study_sample_pcs.to_csv(<span class="st">"study_samples_pcs.txt"</span>, sep<span class="op">=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, index<span class="op">=</span><span class="va">False</span>, </span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a>                        columns<span class="op">=</span>[<span class="st">"FID"</span>, <span class="st">"IID"</span>, <span class="st">"PC1"</span>, <span class="st">"PC2"</span>, <span class="st">"PC3"</span>, <span class="st">"PC4"</span>, <span class="st">"PC5"</span>])</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="adjusting-association-analysis-for-pcs" class="level2" data-number="9.6">
<h2 data-number="9.6" class="anchored" data-anchor-id="adjusting-association-analysis-for-pcs"><span class="header-section-number">9.6</span> Adjusting association analysis for PCs</h2>
<p>Re-run the GWAS of match ticket preference for East Asian teams by executing the following on the Linux command line:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb47"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="ex">~/hpdm208z/programs/plink</span> <span class="dt">\</span></span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">--bfile</span> world_cup_fans <span class="dt">\</span></span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">--1</span> <span class="dt">\</span></span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">--covar</span> study_samples_pcs.txt <span class="dt">\</span></span>
<span id="cb47-5"><a href="#cb47-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">--logistic</span> hide-covar <span class="dt">\</span></span>
<span id="cb47-6"><a href="#cb47-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">--out</span> world_cup_east_asia_team_preference_pc_adjusted</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>Let’s take another look at the number of genetic variants reaching genome-wide level signficance based on a P-value based on 5x10-8 after adjustment for PC values</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb48"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a>world_cup_gwas_results_adjusted <span class="op">=</span> pd.read_csv(<span class="st">"world_cup_east_asia_team_preference_pc_adjusted.assoc.logistic"</span>, sep<span class="op">=</span><span class="vs">r"</span><span class="dv">\s</span><span class="op">+</span><span class="vs">"</span>)</span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a><span class="co"># preview</span></span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(world_cup_gwas_results_adjusted.head())</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>   CHR       SNP      BP A1 TEST  NMISS      OR     STAT        P
0    1   1:14599   14599  A  ADD   1222  1.0160  0.08126  0.93520
1    1   1:14930   14930  G  ADD   1222  0.9182 -0.37020  0.71120
2    1   1:15820   15820  T  ADD   1222  1.0580  0.45270  0.65080
3    1   1:86331   86331  G  ADD   1222  1.5080  2.42900  0.01515
4    1  1:104186  104186  T  ADD   1222  1.2170  1.27900  0.20090</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb50"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Subset to genome-wide significant and order by P-value</span></span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a>world_cup_gwas_results_adjusted_sig <span class="op">=</span> world_cup_gwas_results_adjusted.loc[world_cup_gwas_results_adjusted[<span class="st">'P'</span>] <span class="op">&lt;</span> <span class="fl">5e-08</span>]</span>
<span id="cb50-3"><a href="#cb50-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-4"><a href="#cb50-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Order by P and print top 10 results</span></span>
<span id="cb50-5"><a href="#cb50-5" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(world_cup_gwas_results_adjusted.sort_values(<span class="st">"P"</span>).head())</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>       CHR          SNP         BP A1 TEST  NMISS      OR   STAT         P
17519    7  7:101802456  101802456  A  ADD   1222  2.5100  4.834  0.000001
21629    9  9:103485133  103485133  G  ADD   1222  0.5527 -4.697  0.000003
10808    4  4:143414537  143414537  A  ADD   1222  1.7800  4.523  0.000006
33956   17   17:7864395    7864395  C  ADD   1222  2.1510  4.485  0.000007
34484   17  17:33288655   33288655  G  ADD   1222  0.2253 -4.443  0.000009</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb52"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Count number of GWAS significant</span></span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="bu">len</span>(world_cup_gwas_results_adjusted_sig))</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>0</code></pre>
</div>
</div>
<p>This is what we expected given the phenotype generated for this workshop was for proof-of-principle and made to have zero heritability! It looks like we have adjusted for relative differences in genetic ancestry among individuals in our GWAS analysis.</p>
</section>
</section>
<section id="classifying-genetic-ancestry" class="level1" data-number="10">
<h1 data-number="10"><span class="header-section-number">10</span> Classifying genetic ancestry</h1>
<p>We may want to group individuals into genetic ancestral groups (e.g.&nbsp;AFR, AMR, EUR, EAS, SAS) if we wanted to filter out individuals prior to conducting an analysis e.g.&nbsp;an analysis of individuals only of broad South-Asian genetic ancestry.</p>
<p>There are several supervised machine-learning approaches we can use to do this. For example, we can use a K-nearest neighbour approach to the group individuals into one of several ancestral groups that we have reference data on - which in our case if from the 1000 Genomes Project. To do this we will use a grid search to identify the best combination of values to pass to the parameters of the classifier. Again, we can use Python’s <code>scikit-learn</code> library to do this. First, we will import the required dependencies from library:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb54"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> sklearn.preprocessing <span class="im">import</span> StandardScaler</span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> sklearn.model_selection <span class="im">import</span> GridSearchCV</span>
<span id="cb54-3"><a href="#cb54-3" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> sklearn.neighbors <span class="im">import</span> KNeighborsClassifier</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<section id="defining-targets" class="level2" data-number="10.1">
<h2 data-number="10.1" class="anchored" data-anchor-id="defining-targets"><span class="header-section-number">10.1</span> Defining targets</h2>
<p>Next we will define features (PCs) and targets (genetic ancestry) for samples in the reference data set using the DF created above:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb55"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Standardize the features (PCs) as range of values across PCs differ</span></span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a>standardizer <span class="op">=</span> StandardScaler()                                       <span class="co"># create standardizer</span></span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a>X3 <span class="op">=</span> standardizer.fit_transform(ref_sample_pcs.iloc[:,<span class="dv">2</span>:<span class="dv">7</span>].values)    <span class="co"># standardize first 5 PCs</span></span>
<span id="cb55-4"><a href="#cb55-4" aria-hidden="true" tabindex="-1"></a>y  <span class="op">=</span> ref_sample_pcs.iloc[:,<span class="dv">0</span>].values                                  <span class="co"># define targets</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="creating-and-using-a-grid-search" class="level2" data-number="10.2">
<h2 data-number="10.2" class="anchored" data-anchor-id="creating-and-using-a-grid-search"><span class="header-section-number">10.2</span> Creating and using a grid search</h2>
<p>Set up parameters to test for determining best K-NN parameters to use:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb56"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a>gridParameters <span class="op">=</span> {</span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a>    <span class="st">"n_neighbors"</span> : <span class="bu">range</span>(<span class="dv">3</span>,<span class="dv">100</span>,<span class="dv">2</span>),</span>
<span id="cb56-3"><a href="#cb56-3" aria-hidden="true" tabindex="-1"></a>    <span class="st">"weights"</span>     : [<span class="st">"uniform"</span>, <span class="st">"distance"</span>],</span>
<span id="cb56-4"><a href="#cb56-4" aria-hidden="true" tabindex="-1"></a>    <span class="st">"metric"</span>      : [<span class="st">"euclidean"</span>, <span class="st">"manhattan"</span>, <span class="st">"minkowski"</span>]</span>
<span id="cb56-5"><a href="#cb56-5" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>Run 5-fold cross validation on different K-NN models in the reference data::</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb57"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a>gridSearch        <span class="op">=</span> GridSearchCV(KNeighborsClassifier(), gridParameters, cv<span class="op">=</span><span class="dv">5</span>, verbose<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a>gridSearchResults <span class="op">=</span> gridSearch.fit(X3,y)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>Now we can extract the result of our grid search that resulted in the smallest error on average.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb58"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Print out the results from the grid search of the best combination of parameters to use:</span></span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(gridSearchResults.best_params_)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>{'metric': 'euclidean', 'n_neighbors': 3, 'weights': 'distance'}</code></pre>
</div>
</div>
<p>Python is telling us that the best combination of values to use based on the training data set are:</p>
<ul>
<li>k = 3</li>
<li>distance metric = euclidean</li>
<li>neighbor weighting = distance (i.e.&nbsp;closer neighbors have more weighting)</li>
</ul>
</section>
<section id="assessing-performance-based-on-cross-validation" class="level2" data-number="10.3">
<h2 data-number="10.3" class="anchored" data-anchor-id="assessing-performance-based-on-cross-validation"><span class="header-section-number">10.3</span> Assessing performance based on cross-validation</h2>
<p>Let’s take a look at the confusion matrix based on the best combination of parameters applied to the reference dataset:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb60"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> sklearn.model_selection <span class="im">import</span> cross_val_predict</span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> sklearn.metrics <span class="im">import</span> confusion_matrix</span>
<span id="cb60-3"><a href="#cb60-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-4"><a href="#cb60-4" aria-hidden="true" tabindex="-1"></a>y_pred <span class="op">=</span> cross_val_predict(</span>
<span id="cb60-5"><a href="#cb60-5" aria-hidden="true" tabindex="-1"></a>    gridSearchResults.best_estimator_, X3, y, cv<span class="op">=</span><span class="dv">5</span></span>
<span id="cb60-6"><a href="#cb60-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb60-7"><a href="#cb60-7" aria-hidden="true" tabindex="-1"></a>cm <span class="op">=</span> confusion_matrix(y, y_pred)</span>
<span id="cb60-8"><a href="#cb60-8" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(cm)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>[[317   0   0   0   0]
 [  0 161   0   5   0]
 [  0   0 255   0   0]
 [  0   0   0 232   0]
 [  0   0   0   0 253]]</code></pre>
</div>
</div>
<p>Let’s take a closer look by incorporating data associated with the precision (number or fraction of positives are correct) and recall (number or fraction of positives correctly identified)</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb62"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> sklearn.metrics <span class="im">import</span> classification_report</span>
<span id="cb62-2"><a href="#cb62-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(classification_report(y, y_pred))</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>              precision    recall  f1-score   support

         AFR       1.00      1.00      1.00       317
         AMR       1.00      0.97      0.98       166
         EAS       1.00      1.00      1.00       255
         EUR       0.98      1.00      0.99       232
         SAS       1.00      1.00      1.00       253

    accuracy                           1.00      1223
   macro avg       1.00      0.99      0.99      1223
weighted avg       1.00      1.00      1.00      1223</code></pre>
</div>
</div>
<p>These numbers suggest that 5 individuals who are of Ad-mixed American genetic ancestry have been incorrectly labelled as European but overall this classifier seems to perform well given the confusion matrix, precision and recall.</p>
</section>
<section id="applying-machine-learning-algorithm-with-tuned-hyperparameters" class="level2" data-number="10.4">
<h2 data-number="10.4" class="anchored" data-anchor-id="applying-machine-learning-algorithm-with-tuned-hyperparameters"><span class="header-section-number">10.4</span> Applying machine-learning algorithm with tuned hyperparameters</h2>
<p>Now we have the arguments that minimize the classification error rate in the reference data, we can apply them to our study data to label each individual with their genetic ancestry based on their proximity to the individuals in the reference data. Again, we will use the first 5 PCs to do this.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb64"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create standardized principal components for the study individuals we are trying to classify</span></span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a>X4 <span class="op">=</span> standardizer.transform(study_sample_pcs.iloc[:,<span class="dv">2</span>:<span class="dv">7</span>].values)</span>
<span id="cb64-3"><a href="#cb64-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-4"><a href="#cb64-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Classify and return labels as a list</span></span>
<span id="cb64-5"><a href="#cb64-5" aria-hidden="true" tabindex="-1"></a>study_sample_ancestry_groups <span class="op">=</span> gridSearchResults.predict(X4).tolist() </span>
<span id="cb64-6"><a href="#cb64-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-7"><a href="#cb64-7" aria-hidden="true" tabindex="-1"></a><span class="co"># CREATE "POP" field in the study_samples_pcs DF with the ancestries inferred through KNN </span></span>
<span id="cb64-8"><a href="#cb64-8" aria-hidden="true" tabindex="-1"></a>study_sample_pcs[<span class="st">"POP"</span>] <span class="op">=</span> study_sample_ancestry_groups</span>
<span id="cb64-9"><a href="#cb64-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-10"><a href="#cb64-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Print out the number of individuals classified into each of the populations</span></span>
<span id="cb64-11"><a href="#cb64-11" aria-hidden="true" tabindex="-1"></a>study_sample_pcs[<span class="st">"POP"</span>].value_counts()</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>POP
AFR    309
EUR    270
EAS    242
SAS    226
AMR    175
Name: count, dtype: int64</code></pre>
</div>
</div>
<p>Let’s visually check to see if our KNN classifier has done a seemingly good job by plotting PC1 vs PC2, and PC3 vs PC4. Let’s combine the reference data with the study data and plot all data together:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb66"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Stack the ref PCs and study PCs on top of each other</span></span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a>samples_combined <span class="op">=</span> pd.concat([ref_sample_pcs, study_sample_pcs])</span>
<span id="cb66-3"><a href="#cb66-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb66-4"><a href="#cb66-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Replace missing values created in the "POP" field for the reference data to the FID values that hold the ancestry codes:</span></span>
<span id="cb66-5"><a href="#cb66-5" aria-hidden="true" tabindex="-1"></a>samples_combined[<span class="st">"POP"</span>] <span class="op">=</span> samples_combined[<span class="st">"POP"</span>].fillna(samples_combined[<span class="st">"FID"</span>])</span>
<span id="cb66-6"><a href="#cb66-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb66-7"><a href="#cb66-7" aria-hidden="true" tabindex="-1"></a><span class="co"># PREVIEW</span></span>
<span id="cb66-8"><a href="#cb66-8" aria-hidden="true" tabindex="-1"></a>samples_combined</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>          FID      IID        PC1        PC2        PC3        PC4        PC5  \
0         EUR  HG00097 -14.426319  33.251409  -9.001286   9.522424  -9.248659   
1         EUR  HG00099 -13.755650  34.191333  -9.868827  10.543570 -10.380949   
2         EUR  HG00101 -15.388023  32.814107 -10.876452  11.691985  -6.806517   
3         EUR  HG00102 -14.205595  34.352392  -9.414806  10.879079  -5.199596   
4         EUR  HG00105 -13.501380  34.263998 -10.330081  11.573443  -8.856633   
...       ...      ...        ...        ...        ...        ...        ...   
1217  NA21133  NA21133 -14.312609  10.960837  22.269198  -5.879760   6.244087   
1218  NA21135  NA21135 -14.455674  10.530066  23.284110  -5.778282   6.265334   
1219  NA21141  NA21141 -15.601229  10.736895  22.619389  -3.750600  14.203425   
1220  NA21142  NA21142 -15.895942  10.651730  21.200307  -2.984751  13.984097   
1221  NA21143  NA21143 -16.582568   9.514909  22.259031  -4.312041  12.234487   

           PC6       PC7       PC8       PC9      PC10  POP  
0    -0.878138 -4.520736 -3.365676 -4.221558  1.107865  EUR  
1    -1.693816 -5.966099 -3.783902 -3.527381  0.614973  EUR  
2    -0.587681 -2.750775 -5.732242 -3.818476  2.210309  EUR  
3    -1.299292 -3.553589 -5.761407 -3.533703  0.833023  EUR  
4    -0.963731 -2.920343 -3.335881 -2.428733 -1.363890  EUR  
...        ...       ...       ...       ...       ...  ...  
1217  0.755760 -4.808882 -2.705048  2.953586 -7.807425  SAS  
1218  1.198882 -4.817298 -2.986732  3.898059 -7.458517  SAS  
1219  1.418397 -3.416132  0.227276  1.304175 -7.576578  SAS  
1220  3.393319 -3.191071 -0.544244  0.985557 -8.488138  SAS  
1221  1.617300 -3.690932 -0.241286  2.365615 -7.326827  SAS  

[2445 rows x 13 columns]</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb68"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a>groups <span class="op">=</span> samples_combined.groupby(<span class="st">"POP"</span>)</span>
<span id="cb68-2"><a href="#cb68-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-3"><a href="#cb68-3" aria-hidden="true" tabindex="-1"></a>fig, ax <span class="op">=</span> plt.subplots(figsize <span class="op">=</span> (<span class="dv">10</span>, <span class="dv">8</span>))</span>
<span id="cb68-4"><a href="#cb68-4" aria-hidden="true" tabindex="-1"></a>ax.margins(<span class="fl">0.05</span>)</span>
<span id="cb68-5"><a href="#cb68-5" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> name, group <span class="kw">in</span> groups:</span>
<span id="cb68-6"><a href="#cb68-6" aria-hidden="true" tabindex="-1"></a>    ax.plot(group.PC1, group.PC2, marker<span class="op">=</span><span class="st">"x"</span>, linestyle<span class="op">=</span><span class="st">""</span>, ms<span class="op">=</span><span class="dv">5</span>, label<span class="op">=</span>name)</span>
<span id="cb68-7"><a href="#cb68-7" aria-hidden="true" tabindex="-1"></a>ax.legend()</span>
<span id="cb68-8"><a href="#cb68-8" aria-hidden="true" tabindex="-1"></a>plt.title(<span class="st">"Principal Component Space for Individuals in 1000G Project and Study: PC1 vs PC2"</span>)</span>
<span id="cb68-9"><a href="#cb68-9" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">"PC1"</span>)</span>
<span id="cb68-10"><a href="#cb68-10" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">"PC2"</span>)</span>
<span id="cb68-11"><a href="#cb68-11" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="hpdm208z_workshop8_files/figure-html/unnamed-chunk-40-7.png" class="img-fluid figure-img" width="960"></p>
</figure>
</div>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb69"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a>fig, ax <span class="op">=</span> plt.subplots(figsize <span class="op">=</span> (<span class="dv">10</span>, <span class="dv">8</span>))</span>
<span id="cb69-2"><a href="#cb69-2" aria-hidden="true" tabindex="-1"></a>ax.margins(<span class="fl">0.05</span>)</span>
<span id="cb69-3"><a href="#cb69-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> name, group <span class="kw">in</span> groups:</span>
<span id="cb69-4"><a href="#cb69-4" aria-hidden="true" tabindex="-1"></a>    ax.plot(group.PC3, group.PC4, marker<span class="op">=</span><span class="st">"x"</span>, linestyle<span class="op">=</span><span class="st">""</span>, ms<span class="op">=</span><span class="dv">5</span>, label<span class="op">=</span>name)</span>
<span id="cb69-5"><a href="#cb69-5" aria-hidden="true" tabindex="-1"></a>ax.legend()</span>
<span id="cb69-6"><a href="#cb69-6" aria-hidden="true" tabindex="-1"></a>plt.title(<span class="st">"Principal Component Space for Individuals in 1000G Project and Study: PC3 vs PC4"</span>)</span>
<span id="cb69-7"><a href="#cb69-7" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">"PC3"</span>)</span>
<span id="cb69-8"><a href="#cb69-8" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">"PC4"</span>)</span>
<span id="cb69-9"><a href="#cb69-9" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="hpdm208z_workshop8_files/figure-html/unnamed-chunk-40-8.png" class="img-fluid figure-img" width="960"></p>
</figure>
</div>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb70"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a>fig, ax <span class="op">=</span> plt.subplots(figsize <span class="op">=</span> (<span class="dv">10</span>, <span class="dv">8</span>))</span>
<span id="cb70-2"><a href="#cb70-2" aria-hidden="true" tabindex="-1"></a>ax.margins(<span class="fl">0.05</span>)</span>
<span id="cb70-3"><a href="#cb70-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> name, group <span class="kw">in</span> groups:</span>
<span id="cb70-4"><a href="#cb70-4" aria-hidden="true" tabindex="-1"></a>    ax.plot(group.PC5, group.PC6, marker<span class="op">=</span><span class="st">"x"</span>, linestyle<span class="op">=</span><span class="st">""</span>, ms<span class="op">=</span><span class="dv">5</span>, label<span class="op">=</span>name)</span>
<span id="cb70-5"><a href="#cb70-5" aria-hidden="true" tabindex="-1"></a>ax.legend()</span>
<span id="cb70-6"><a href="#cb70-6" aria-hidden="true" tabindex="-1"></a>plt.title(<span class="st">"Principal Component Space for Individuals in 1000G Project and Study: PC5 vs PC6"</span>)</span>
<span id="cb70-7"><a href="#cb70-7" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">"PC5"</span>)</span>
<span id="cb70-8"><a href="#cb70-8" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">"PC6"</span>)</span>
<span id="cb70-9"><a href="#cb70-9" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="hpdm208z_workshop8_files/figure-html/unnamed-chunk-40-9.png" class="img-fluid figure-img" width="960"></p>
</figure>
</div>
</div>
</div>
<p>It looks like our classifier has done a reasonable job at classifying individuals into the 5 broad genetic ancestries in our reference data set.</p>
</section>
<section id="task-2" class="level2" data-number="10.5">
<h2 data-number="10.5" class="anchored" data-anchor-id="task-2"><span class="header-section-number">10.5</span> Task</h2>
<ul>
<li>A mapping of 1000 Genomes Project identifiers and their the sub-populations can be found on the ELE page. As an extension to this workshop, see if you can re-label both the reference samples and study samples with these more refined ancestries.</li>
</ul>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
      const outerScaffold = trigger.parentElement.cloneNode(true);
      const codeEl = outerScaffold.querySelector('code');
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




</body></html>