<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.8.25">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Prof.&nbsp;Andrew R Wood">
<meta name="dcterms.date" content="2026-01-26">

<title>HPDM208Z Stratified Medicine Workshop 9: Genetic predictors of human health</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="hpdm208z_workshop9_files/libs/clipboard/clipboard.min.js"></script>
<script src="hpdm208z_workshop9_files/libs/quarto-html/quarto.js" type="module"></script>
<script src="hpdm208z_workshop9_files/libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="hpdm208z_workshop9_files/libs/quarto-html/axe/axe-check.js" type="module"></script>
<script src="hpdm208z_workshop9_files/libs/quarto-html/popper.min.js"></script>
<script src="hpdm208z_workshop9_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="hpdm208z_workshop9_files/libs/quarto-html/anchor.min.js"></script>
<link href="hpdm208z_workshop9_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="hpdm208z_workshop9_files/libs/quarto-html/quarto-syntax-highlighting-7b89279ff1a6dce999919e0e67d4d9ec.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="hpdm208z_workshop9_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="hpdm208z_workshop9_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="hpdm208z_workshop9_files/libs/bootstrap/bootstrap-d5214d16172734abcd1dcccc621ce81b.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN" && texText && texText.data) {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

</head>

<body class="quarto-light">

<header id="title-block-header" class="quarto-title-block default page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <h1 class="title">HPDM208Z Stratified Medicine Workshop 9: Genetic predictors of human health</h1>
            <p class="subtitle lead">MSc Health Data Science, University of Exeter</p>
                      </div>
  </div>
    
  
  <div class="quarto-title-meta">

      <div>
      <div class="quarto-title-meta-heading">Author</div>
      <div class="quarto-title-meta-contents">
               <p>Prof.&nbsp;Andrew R Wood <a href="mailto:A.R.Wood@exeter.ac.uk" class="quarto-title-author-email"><i class="bi bi-envelope"></i></a> </p>
            </div>
    </div>
      
      <div>
      <div class="quarto-title-meta-heading">Published</div>
      <div class="quarto-title-meta-contents">
        <p class="date">2026-01-26</p>
      </div>
    </div>
    
      
    </div>
    
  
  </header><div id="quarto-content" class="page-columns page-rows-contents page-layout-article">
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
  <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction"><span class="header-section-number">1</span> Introduction</a></li>
  <li><a href="#aims-of-workshop" id="toc-aims-of-workshop" class="nav-link" data-scroll-target="#aims-of-workshop"><span class="header-section-number">2</span> Aims of workshop</a></li>
  <li><a href="#conda-environment" id="toc-conda-environment" class="nav-link" data-scroll-target="#conda-environment"><span class="header-section-number">3</span> Conda environment</a></li>
  <li><a href="#workshop-directory" id="toc-workshop-directory" class="nav-link" data-scroll-target="#workshop-directory"><span class="header-section-number">4</span> Workshop directory</a></li>
  <li><a href="#workshop-data" id="toc-workshop-data" class="nav-link" data-scroll-target="#workshop-data"><span class="header-section-number">5</span> Workshop data</a></li>
  <li><a href="#generating-a-prs-for-coronary-artery-disease" id="toc-generating-a-prs-for-coronary-artery-disease" class="nav-link" data-scroll-target="#generating-a-prs-for-coronary-artery-disease"><span class="header-section-number">6</span> Generating a PRS for Coronary Artery Disease</a>
  <ul class="collapse">
  <li><a href="#aligning-allele-counts-to-disease-risk-increasing-alleles" id="toc-aligning-allele-counts-to-disease-risk-increasing-alleles" class="nav-link" data-scroll-target="#aligning-allele-counts-to-disease-risk-increasing-alleles"><span class="header-section-number">6.1</span> Aligning allele counts to disease-risk increasing alleles</a></li>
  <li><a href="#weighting-numbers-of-disease-risk-increasing-allele-by-their-effects" id="toc-weighting-numbers-of-disease-risk-increasing-allele-by-their-effects" class="nav-link" data-scroll-target="#weighting-numbers-of-disease-risk-increasing-allele-by-their-effects"><span class="header-section-number">6.2</span> Weighting numbers of disease risk increasing allele by their effects</a></li>
  <li><a href="#comparing-the-distribution-of-cad-prs-between-cases-and-controls" id="toc-comparing-the-distribution-of-cad-prs-between-cases-and-controls" class="nav-link" data-scroll-target="#comparing-the-distribution-of-cad-prs-between-cases-and-controls"><span class="header-section-number">6.3</span> Comparing the distribution of CAD-PRS between cases and controls</a></li>
  <li><a href="#questions" id="toc-questions" class="nav-link" data-scroll-target="#questions"><span class="header-section-number">6.4</span> Questions</a></li>
  <li><a href="#comparing-prevalence-of-cases-across-the-cad-prs-distribution" id="toc-comparing-prevalence-of-cases-across-the-cad-prs-distribution" class="nav-link" data-scroll-target="#comparing-prevalence-of-cases-across-the-cad-prs-distribution"><span class="header-section-number">6.5</span> Comparing prevalence of cases across the CAD-PRS distribution</a></li>
  <li><a href="#assessing-the-overall-predictive-accuracy-of-the-cad-prs" id="toc-assessing-the-overall-predictive-accuracy-of-the-cad-prs" class="nav-link" data-scroll-target="#assessing-the-overall-predictive-accuracy-of-the-cad-prs"><span class="header-section-number">6.6</span> Assessing the overall predictive accuracy of the CAD-PRS</a></li>
  <li><a href="#question" id="toc-question" class="nav-link" data-scroll-target="#question"><span class="header-section-number">6.7</span> Question</a></li>
  </ul></li>
  <li><a href="#task---type-2-diabetes" id="toc-task---type-2-diabetes" class="nav-link" data-scroll-target="#task---type-2-diabetes"><span class="header-section-number">7</span> Task - Type 2 diabetes</a></li>
  <li><a href="#incorporting-more-features-snps-into-a-prs" id="toc-incorporting-more-features-snps-into-a-prs" class="nav-link" data-scroll-target="#incorporting-more-features-snps-into-a-prs"><span class="header-section-number">8</span> Incorporting more features (SNPs) into a PRS</a></li>
  <li><a href="#generating-several-prs-variables" id="toc-generating-several-prs-variables" class="nav-link" data-scroll-target="#generating-several-prs-variables"><span class="header-section-number">9</span> Generating several PRS variables</a>
  <ul class="collapse">
  <li><a href="#questions-1" id="toc-questions-1" class="nav-link" data-scroll-target="#questions-1"><span class="header-section-number">9.1</span> Questions:</a></li>
  </ul></li>
  <li><a href="#taking-the-p-value-threshold-forwards-into-an-independent-study" id="toc-taking-the-p-value-threshold-forwards-into-an-independent-study" class="nav-link" data-scroll-target="#taking-the-p-value-threshold-forwards-into-an-independent-study"><span class="header-section-number">10</span> Taking the P-value threshold forwards into an independent study</a>
  <ul class="collapse">
  <li><a href="#questions-and-tasks" id="toc-questions-and-tasks" class="nav-link" data-scroll-target="#questions-and-tasks"><span class="header-section-number">10.1</span> Questions and tasks</a></li>
  </ul></li>
  </ul>
</nav>
</div>
<main class="content quarto-banner-title-block" id="quarto-document-content">





<section id="introduction" class="level1" data-number="1">
<h1 data-number="1"><span class="header-section-number">1</span> Introduction</h1>
<p>In this workshop, you will learn how to construct and evaluate polygenic risk scores (PRS) as genetic predictors of human health. You will generate PRSs for coronary artery disease (CAD) and type 2 diabetes (T2D).</p>
</section>
<section id="aims-of-workshop" class="level1" data-number="2">
<h1 data-number="2"><span class="header-section-number">2</span> Aims of workshop</h1>
<ul>
<li>enable you to generate a PRS</li>
<li>compare frequency of cases and controls across quantiles for a PRS distribution</li>
<li>assess the overall predictive accuracy of a PRS in determining disease status</li>
<li>enable you to generate a PRS at different P-value thresholds</li>
<li>compare ROC-AUCs to determine strongest PRS based on different P-value thresholds</li>
<li>assess the overall predictive accuracy of the PRS in determining disease status in an independent study</li>
</ul>
</section>
<section id="conda-environment" class="level1" data-number="3">
<h1 data-number="3"><span class="header-section-number">3</span> Conda environment</h1>
<p>Before attempting this workshop, you should activate the <code>hpdm208z</code> conda environment that will provide the R through Jupyter Lab and required library dependencies:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate hpdm208z</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="workshop-directory" class="level1" data-number="4">
<h1 data-number="4"><span class="header-section-number">4</span> Workshop directory</h1>
<p>All relevant files associated with this workshops can be found in the following directory:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb2"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="ex">/home/ubuntu/hpdm208z/workshops/prs/</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>You should navigate to this directory prior to starting the workshop. Note, a jupyter notebook can be be loaded from this directory to assist you in undertaking this workshop:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb3"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="ex">prs.ipynb</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="workshop-data" class="level1" data-number="5">
<h1 data-number="5"><span class="header-section-number">5</span> Workshop data</h1>
<p>Below is a table describing the data files in the workshops directory:</p>
<table class="caption-top table">
<colgroup>
<col style="width: 30%">
<col style="width: 70%">
</colgroup>
<thead>
<tr class="header">
<th>File Name</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>cad_phenotype.txt</code></td>
<td>CAD case-control status on 9,766 individuals</td>
</tr>
<tr class="even">
<td><code>cad_snp_genotypes.txt</code></td>
<td>Genotype data on 66 SNPs previously associated with CAD in 9,766 individuals</td>
</tr>
<tr class="odd">
<td><code>cad_snp_weights.txt</code></td>
<td>Allele effect estimates of 66 SNPs previously associated with CAD</td>
</tr>
<tr class="even">
<td><code>t2d_phenotype.txt</code></td>
<td>T2D case-control status on 9,755 individuals</td>
</tr>
<tr class="odd">
<td><code>t2d_snp_genotypes.txt</code></td>
<td>Genotype data on 88 SNPs previously associated with T2D in 9,755 individuals</td>
</tr>
<tr class="even">
<td><code>t2d_snp_weights.txt</code></td>
<td>Allele effect estimates of 88 SNPs previously associated with T2D</td>
</tr>
<tr class="odd">
<td></td>
<td></td>
</tr>
<tr class="even">
<td><code>CAD_GWAS.txt</code></td>
<td>Published GWAS results for CAD (Nikpay <em>et al.</em> 2015, Nature Genetics)</td>
</tr>
<tr class="odd">
<td><code>study1_snp_genotypes.txt</code></td>
<td>Genotype data on 7,114 SNPs in 10,000 individuals</td>
</tr>
<tr class="even">
<td><code>study1_cad_status.txt</code></td>
<td>CAD case-control status on 10,000 individuals</td>
</tr>
<tr class="odd">
<td><code>study2_snp_genotypes.txt</code></td>
<td>Genotype data on 7,114 SNPs in 9,766 individuals</td>
</tr>
<tr class="even">
<td><code>study2_cad_status.txt</code></td>
<td>CAD case-control status on 9,766 individuals</td>
</tr>
</tbody>
</table>
<hr>
</section>
<section id="generating-a-prs-for-coronary-artery-disease" class="level1" data-number="6">
<h1 data-number="6"><span class="header-section-number">6</span> Generating a PRS for Coronary Artery Disease</h1>
<p>Let’s first load in genotype data for 66 SNPs known to be associated with Coronary Artery Disease (CAD) in 9,000 individuals using the file <code>cad_snp_genotypes.txt</code> into a DF called <code>cad_snp_genotypes</code>:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Read data in:</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>cad_snp_genotypes <span class="ot">=</span> <span class="fu">read.table</span>(<span class="st">"cad_snp_genotypes.txt"</span>, <span class="at">header=</span>T, <span class="at">stringsAsFactors=</span>F)</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Preview first 10 rows and columns:</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>cad_snp_genotypes[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>         Rsid Allele1 Allele2 P7212 P3362 P4481 P1345 P4276 P89 P4066
1   rs9970807       C       T     1     1     1     0     0   0     0
2   rs7528419       A       G     0     1     1     1     0   2     0
3   rs6689306       A       G     2     1     0     0     2   1     1
4   rs1892094       C       T     2     0     1     0     1   0     1
5   rs6700559       C       T     0     0     2     2     2   0     1
6   rs2820315       C       T     1     1     0     2     0   0     0
7  rs67180937       T       G     0     2     1     2     2   2     2
8  rs16986953       G       A     0     0     1     1     1   0     1
9   rs7568458       T       A     1     0     1     1     0   2     1
10 rs17678683       T       G     0     0     0     1     0   0     0</code></pre>
</div>
</div>
<p>Note that the rows refer to a specific variant, while the 3rd column onwards are specific to individuals. Each column associated with an individual contains numerical representations of genotypes for the SNPs associated with CAD: <code>[0,1,2]</code>. The numerical genotype codes refer to the number of copies of <code>Allele2</code> carried by each individual.</p>
<p>Now we will load data on the SNPs associated with CAD-risk held in <code>cad_snp_weights.txt</code> into a DF called <code>cad_snp_weights</code>. This data includes</p>
<ul>
<li>the disease risk increasing allele (<code>Effect_Allele</code> in this dataset)</li>
<li>the associated effect as ln(OR) per <code>Effect_Allele</code> allele carried <code>Weight</code> in this dataset).</li>
</ul>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Read data in:</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>cad_snp_weights <span class="ot">=</span> <span class="fu">read.table</span>(<span class="st">"cad_snp_weights.txt"</span>, <span class="at">header=</span>T, <span class="at">stringsAsFactors=</span>F)</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Preview first 10 rows and columns:</span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(cad_snp_weights)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>        Rsid Risk_Allele Other_Allele     Weight
1  rs2891168           G            A 0.19062036
2  rs9349379           G            A 0.13102826
3 rs55730499           T            C 0.31481074
4  rs2315065           A            C 0.28517894
5  rs7528419           A            G 0.11332868
6  rs4468572           C            T 0.07696104</code></pre>
</div>
</div>
<p>We can align the information about SNP effects with genotypes for those SNPs by merging the above DFs together. We will create a DF called <code>cad_genetics</code> that provides the merged data of CAD-SNP genotypes and weights.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Read data in:</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>cad_genetics <span class="ot">=</span> <span class="fu">merge</span>(cad_snp_weights, cad_snp_genotypes, <span class="at">by=</span><span class="st">"Rsid"</span>)</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Preview first 10 rows and columns:</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>cad_genetics[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>         Rsid Risk_Allele Other_Allele     Weight Allele1 Allele2 P7212 P3362 P4481 P1345
1  rs10139550           G            C 0.05826891       C       G     1     0     0     1
2  rs10237377           T            G 0.05129329       G       T     1     1     1     0
3   rs1050362           A            C 0.03922071       C       A     1     1     1     1
4  rs10840293           A            G 0.05826891       G       A     1     1     1     0
5  rs11057830           A            G 0.06765865       G       A     1     1     1     0
6  rs11065979           T            C 0.06765865       C       T     1     1     2     0
7  rs11191416           T            G 0.07696104       T       G     0     0     0     0
8  rs11556924           C            T 0.07696104       C       T     1     1     1     2
9  rs11617955           T            A 0.08617770       T       A     0     1     0     0
10 rs11838776           A            G 0.06765865       G       A     1     1     0     1</code></pre>
</div>
</div>
<section id="aligning-allele-counts-to-disease-risk-increasing-alleles" class="level2" data-number="6.1">
<h2 data-number="6.1" class="anchored" data-anchor-id="aligning-allele-counts-to-disease-risk-increasing-alleles"><span class="header-section-number">6.1</span> Aligning allele counts to disease-risk increasing alleles</h2>
<p>When creating a PRS, we typically align genotypes to represent the number of risk increasing alleles carried (or number of trait raising alleles for continuous traits). However, the allele that represents the increasing genotype code for a given dataset is usually randomized to what may a disease-specific risk-increasing allele. In the DFs above, the risk allele is defined in the <code>Risk_Allele</code> column and the genotype codes are aligned to <code>Allele2</code>. So we need to align the genotype codes <code>[0,1,2]</code> to represent the number of disease-risk increasing alleles carried where required (i.e.&nbsp;<code>Risk_Allele</code> != <code>Allele2</code>) . We will later weight the aligned genotype codes by the effect (logOR) of the disease risk raising allele.</p>
<p>The following code will identify rows where <code>Risk_Allele</code> != <code>Allele2</code> and process values in the 7th column to the last column to flip the genotype codes so that <code>0 -&gt; 2</code> and <code>2 -&gt; 0</code>. Values of <code>1</code> can remain the same because they represent 1 copy of the <code>Risk_Allele</code> and 1 copy of the <code>Other_Allele</code>.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>cad_genetics[cad_genetics<span class="sc">$</span>Risk_Allele <span class="sc">!=</span> cad_genetics<span class="sc">$</span>Allele2, <span class="dv">7</span><span class="sc">:</span><span class="fu">ncol</span>(cad_genetics)] <span class="ot">=</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>   <span class="dv">2</span> <span class="sc">-</span> cad_genetics[cad_genetics<span class="sc">$</span>Risk_Allele <span class="sc">!=</span> cad_genetics<span class="sc">$</span>Allele2, <span class="dv">7</span><span class="sc">:</span><span class="fu">ncol</span>(cad_genetics)] </span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Preview first 10 rows and columns:</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>cad_genetics[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>         Rsid Risk_Allele Other_Allele     Weight Allele1 Allele2 P7212 P3362 P4481 P1345
1  rs10139550           G            C 0.05826891       C       G     1     0     0     1
2  rs10237377           T            G 0.05129329       G       T     1     1     1     0
3   rs1050362           A            C 0.03922071       C       A     1     1     1     1
4  rs10840293           A            G 0.05826891       G       A     1     1     1     0
5  rs11057830           A            G 0.06765865       G       A     1     1     1     0
6  rs11065979           T            C 0.06765865       C       T     1     1     2     0
7  rs11191416           T            G 0.07696104       T       G     2     2     2     2
8  rs11556924           C            T 0.07696104       C       T     1     1     1     0
9  rs11617955           T            A 0.08617770       T       A     2     1     2     2
10 rs11838776           A            G 0.06765865       G       A     1     1     0     1</code></pre>
</div>
</div>
<p>By previewing the <code>cad_genetics</code> DF, we can see that the genotype codes remain the same for the first 6 SNPs but the 7th SNP <code>rs11191416</code> has the respective genotypes flipped because the coding was not originally aligned to the disease risk increasing allele. This is also true for the next 3 SNPs in the data preview.</p>
<p>For are now going to create a copy of the <code>cad_genetics</code> DF and modify the aligned genotype codes to be weighted by the effect of each respective genetic variant. We will do this by multiplying the genotype code to the logOR held in the <code>Weight</code> column. We will first create a DF called <code>cad_genetics_weighted</code>:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Copy DF containing genetic data related to CAD</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>cad_genetics_weighted <span class="ot">=</span> cad_genetics</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="weighting-numbers-of-disease-risk-increasing-allele-by-their-effects" class="level2" data-number="6.2">
<h2 data-number="6.2" class="anchored" data-anchor-id="weighting-numbers-of-disease-risk-increasing-allele-by-their-effects"><span class="header-section-number">6.2</span> Weighting numbers of disease risk increasing allele by their effects</h2>
<p>Now we will perform the multiplication of genotype values by the respecitive effect of the variant on CAD risk.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Copy DF containing genetic data related to CAD</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>cad_genetics_weighted[<span class="dv">7</span><span class="sc">:</span><span class="fu">ncol</span>(cad_genetics_weighted)] <span class="ot">=</span> cad_genetics_weighted[,<span class="dv">7</span><span class="sc">:</span><span class="fu">ncol</span>(cad_genetics_weighted)] <span class="sc">*</span> cad_genetics_weighted<span class="sc">$</span>Weight</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Preview first 10 rows and columns:</span></span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>cad_genetics_weighted[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>         Rsid Risk_Allele Other_Allele     Weight Allele1 Allele2      P7212      P3362      P4481      P1345
1  rs10139550           G            C 0.05826891       C       G 0.05826891 0.00000000 0.00000000 0.05826891
2  rs10237377           T            G 0.05129329       G       T 0.05129329 0.05129329 0.05129329 0.00000000
3   rs1050362           A            C 0.03922071       C       A 0.03922071 0.03922071 0.03922071 0.03922071
4  rs10840293           A            G 0.05826891       G       A 0.05826891 0.05826891 0.05826891 0.00000000
5  rs11057830           A            G 0.06765865       G       A 0.06765865 0.06765865 0.06765865 0.00000000
6  rs11065979           T            C 0.06765865       C       T 0.06765865 0.06765865 0.13531730 0.00000000
7  rs11191416           T            G 0.07696104       T       G 0.15392208 0.15392208 0.15392208 0.15392208
8  rs11556924           C            T 0.07696104       C       T 0.07696104 0.07696104 0.07696104 0.00000000
9  rs11617955           T            A 0.08617770       T       A 0.17235539 0.08617770 0.17235539 0.17235539
10 rs11838776           A            G 0.06765865       G       A 0.06765865 0.06765865 0.00000000 0.06765865</code></pre>
</div>
</div>
<p>Finally, we can create a DF that contains subjects IDs and their respective CAD PRS by summing up the weighted alleles for individual:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Copy DF containing genetic data related to CAD</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>cad_prs <span class="ot">=</span> <span class="fu">data.frame</span>(</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">id =</span> <span class="fu">names</span>(cad_genetics_weighted)[<span class="dv">7</span><span class="sc">:</span><span class="fu">ncol</span>(cad_genetics_weighted)],</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">prs =</span> <span class="fu">colSums</span>(cad_genetics_weighted[<span class="dv">7</span><span class="sc">:</span><span class="fu">ncol</span>(cad_genetics_weighted)]), </span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">row.names =</span> <span class="cn">NULL</span></span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a><span class="co"># preview</span></span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(cad_prs)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>     id      prs
1 P7212 4.381356
2 P3362 3.762699
3 P4481 4.345376
4 P1345 4.507998
5 P4276 4.503723
6   P89 3.776389</code></pre>
</div>
</div>
<p>At this point we have a PRS for each individual which provides a summary statistic of their overall genetic susceptibility for CAD. The value on its own is of little use and should be combined with other data in order to perform meaningful analyses. For example, we may want visualise the distribution of CAD PRS among cases and controls, know what the prevalence of CAD is among different quantiles of the PRS distribution, or get a sense of whether we could classify a patient who may have symptoms of CAD using the PRS without adding any additional information into the model through the use of ROC curves.</p>
<hr>
</section>
<section id="comparing-the-distribution-of-cad-prs-between-cases-and-controls" class="level2" data-number="6.3">
<h2 data-number="6.3" class="anchored" data-anchor-id="comparing-the-distribution-of-cad-prs-between-cases-and-controls"><span class="header-section-number">6.3</span> Comparing the distribution of CAD-PRS between cases and controls</h2>
<p>We will now compare how the PRS values compare between cases and controls for CAD. Intuitively, we should expect CAD cases to have higher PRS values on average than CAD controls so let’s check. First we will load in the phenotype file containing case (<code>1</code>) and control (<code>0</code>) status.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load in the phenotype data into a DF</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>cad_pheno <span class="ot">=</span> <span class="fu">read.table</span>(<span class="st">"cad_phenotype.txt"</span>, <span class="at">header=</span>T, <span class="at">stringsAsFactors =</span> F)</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Preview DF</span></span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(cad_pheno)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>  id cad_status
1 P1          0
2 P2          0
3 P3          0
4 P4          1
5 P5          0
6 P6          0</code></pre>
</div>
</div>
<p>Now we will merge the data held in the <code>cad_prs</code> DF.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load in the phenotype data into a DF</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>cad_pheno_and_prs <span class="ot">=</span> <span class="fu">merge</span>(cad_pheno, cad_prs, <span class="at">by=</span><span class="st">"id"</span>)</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Preview DF</span></span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(cad_pheno_and_prs)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>     id cad_status      prs
1    P1          0 4.008102
2   P10          0 4.096070
3  P100          1 5.123279
4 P1000          0 4.871205
5 P1001          0 4.163969
6 P1002          0 4.876188</code></pre>
</div>
</div>
<p>Let’s plot the distribution of the GRS in cases and controls, overlaying the two distributions into a single histogram with transparency (i.e.&nbsp;set alpha color to 0.5). Use 100 bins when plotting the data.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>  cad_pheno_and_prs, <span class="fu">aes</span>(<span class="at">x=</span>prs, <span class="at">fill =</span> <span class="fu">as.factor</span>(cad_status))) <span class="sc">+</span> </span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">position =</span> <span class="st">"identity"</span>, <span class="at">alpha =</span> <span class="fl">0.5</span>, <span class="at">bins =</span> <span class="dv">30</span>) <span class="sc">+</span> </span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(</span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> <span class="st">"CAD PRS"</span>,</span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> <span class="st">"N individuals"</span>,</span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">fill =</span> <span class="st">"Case / control status"</span></span>
<span id="cb21-10"><a href="#cb21-10" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="hpdm208z_workshop9_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="questions" class="level2" data-number="6.4">
<h2 data-number="6.4" class="anchored" data-anchor-id="questions"><span class="header-section-number">6.4</span> Questions</h2>
<ul>
<li>What do you notice about the distribution of PRS values among cases compared to controls?</li>
<li>Based on the histogram, do you think observed differences make the PRS a good diagnostic tool for the disease?</li>
</ul>
<p><strong>Return to the course page to post your thoughts</strong></p>
<hr>
</section>
<section id="comparing-prevalence-of-cases-across-the-cad-prs-distribution" class="level2" data-number="6.5">
<h2 data-number="6.5" class="anchored" data-anchor-id="comparing-prevalence-of-cases-across-the-cad-prs-distribution"><span class="header-section-number">6.5</span> Comparing prevalence of cases across the CAD-PRS distribution</h2>
<p>We will now compare the frequency of CAD cases among different quartiles of the CAD-PRS distribution.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Get quintile vales from CAD PRS distribution:</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>quartiles <span class="ot">&lt;-</span> <span class="fu">quantile</span>(cad_pheno_and_prs<span class="sc">$</span>prs, <span class="at">probs =</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="fl">0.25</span>))</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a><span class="co"># create a respective list of labels we can assign PRS values to</span></span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a>quartile_labels <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Q1"</span>, <span class="st">"Q2"</span>, <span class="st">"Q3"</span>, <span class="st">"Q4"</span>)</span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-7"><a href="#cb22-7" aria-hidden="true" tabindex="-1"></a><span class="co"># create a new column in cad_pheno_and_prs DF which labels the quintile the PRS falls into</span></span>
<span id="cb22-8"><a href="#cb22-8" aria-hidden="true" tabindex="-1"></a>cad_pheno_and_prs<span class="sc">$</span>prs_quartile <span class="ot">&lt;-</span> <span class="fu">cut</span>(</span>
<span id="cb22-9"><a href="#cb22-9" aria-hidden="true" tabindex="-1"></a>  cad_pheno_and_prs<span class="sc">$</span>prs,</span>
<span id="cb22-10"><a href="#cb22-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">breaks =</span> quartiles,</span>
<span id="cb22-11"><a href="#cb22-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">include.lowest =</span> <span class="cn">TRUE</span>,  <span class="co"># include the smallest value</span></span>
<span id="cb22-12"><a href="#cb22-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">labels =</span> quartile_labels</span>
<span id="cb22-13"><a href="#cb22-13" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb22-14"><a href="#cb22-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-15"><a href="#cb22-15" aria-hidden="true" tabindex="-1"></a><span class="co"># preview df</span></span>
<span id="cb22-16"><a href="#cb22-16" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(cad_pheno_and_prs)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>     id cad_status      prs prs_quartile
1    P1          0 4.008102           Q1
2   P10          0 4.096070           Q2
3  P100          1 5.123279           Q4
4 P1000          0 4.871205           Q4
5 P1001          0 4.163969           Q2
6 P1002          0 4.876188           Q4</code></pre>
</div>
</div>
<p>Now we have determined the quartiles the PRS values fall into, we can tabulate case control status</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>counts <span class="ot">=</span> <span class="fu">table</span>(cad_pheno_and_prs<span class="sc">$</span>prs_quartile, cad_pheno_and_prs<span class="sc">$</span>cad_status)</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>props <span class="ot">=</span> <span class="fu">prop.table</span>(counts, <span class="at">margin =</span> <span class="dv">1</span>)</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a><span class="co"># print counts of cases (1) and controls (0) per PRS quartile</span></span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(counts)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>    
        0    1
  Q1 1369 1073
  Q2 1265 1176
  Q3 1215 1226
  Q4 1020 1422</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="co"># print proportions of cases (1) and controls (0) per PRS quartile to 4dp</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">round</span>(props, <span class="dv">4</span>))</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>    
          0      1
  Q1 0.5606 0.4394
  Q2 0.5182 0.4818
  Q3 0.4977 0.5023
  Q4 0.4177 0.5823</code></pre>
</div>
</div>
<p>Clearly the frequency of cases is increasing as we are moving up through the quartiles of the GRS. You may want to play around with the data and split the data into quintiles or deciles and compare what happens to the frequency of cases in those groups.</p>
<hr>
</section>
<section id="assessing-the-overall-predictive-accuracy-of-the-cad-prs" class="level2" data-number="6.6">
<h2 data-number="6.6" class="anchored" data-anchor-id="assessing-the-overall-predictive-accuracy-of-the-cad-prs"><span class="header-section-number">6.6</span> Assessing the overall predictive accuracy of the CAD-PRS</h2>
<p>We may have a new patient with a condition that is undiagnosed but there may a chance they have CAD. We could try to predict their CAD status based on our PRS but to do that we will need to create a ROC curve and calculate the area under the curve (AUC) to determine how good a predictor it will be! We will use the <code>pROC</code> R library to do this.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressWarnings</span>(<span class="fu">suppressPackageStartupMessages</span>(<span class="fu">library</span>(pROC)))</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a><span class="co"># create roc object we can use for funtions associated with pROC</span></span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>roc_obj <span class="ot">=</span> <span class="fu">roc</span>(cad_pheno_and_prs<span class="sc">$</span>cad_status , cad_pheno_and_prs<span class="sc">$</span>prs)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Setting levels: control = 0, case = 1</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Setting direction: controls &lt; cases</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(roc_obj, <span class="at">col =</span> <span class="st">"blue"</span>, <span class="at">lwd =</span> <span class="dv">2</span>, <span class="at">main =</span> <span class="st">"ROC Curve"</span>, <span class="at">legacy.axes =</span> <span class="cn">TRUE</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="hpdm208z_workshop9_files/figure-html/unnamed-chunk-17-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate the area under the curve:</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>auc_value <span class="ot">=</span> <span class="fu">auc</span>(roc_obj)</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(auc_value)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>Area under the curve: 0.5626</code></pre>
</div>
</div>
</section>
<section id="question" class="level2" data-number="6.7">
<h2 data-number="6.7" class="anchored" data-anchor-id="question"><span class="header-section-number">6.7</span> Question</h2>
<ul>
<li>Does the ROC-AUC suggest the PRS would be a good diagnostic tool for CAD in a clinical setting?</li>
</ul>
<p><strong>Return to the course page to post your thoughts</strong></p>
<hr>
</section>
</section>
<section id="task---type-2-diabetes" class="level1" data-number="7">
<h1 data-number="7"><span class="header-section-number">7</span> Task - Type 2 diabetes</h1>
<ul>
<li>Repeat the steps above for type 2 diabetes and answer the same questions again. You will need the following files:
<ul>
<li><code>t2d_phenotype.txt</code></li>
<li><code>t2d_snp_genotypes.txt</code></li>
<li><code>t2d_snp_weights.txt</code></li>
</ul></li>
<li>Compare the ROC-AUC from the T2D PRS for type 2 diabetes with the ROC-AUC from the CAD PRS for coronary artery disease. Do you know notice any differences?</li>
</ul>
<p><strong>Return to the course page to post your thoughts</strong></p>
</section>
<section id="incorporting-more-features-snps-into-a-prs" class="level1" data-number="8">
<h1 data-number="8"><span class="header-section-number">8</span> Incorporting more features (SNPs) into a PRS</h1>
<p>So far, you have generated a PRS for coronary artery disease (CAD) and type 2 diabetes (T2D) using genetic variants discovered to be associated with these traits in independent studies at P&lt;5x10-8. You checked for differences in prevalence across quartiles of the disease-specific PRS and examined the area under the ROC curve to assess overall predictive accuracy of a genetic predictor.</p>
<p>We will now do something similar except you will be incorporating more data from across the genome that may not have met our threshold of statistical significance. A research study aiming to identify associations will always have a limit in the magnitude of effect sizes it is able to detected at a given level of statistical significance due to its sample size. It’s possible that more genetic variation could be included in the predictor but the effects are too subtle to be detected by a study owing to it’s sample size. If a study had access to a larger sample size, it is possible that smaller effects could be detected. Recent work in the field has shown that by including more genetic variants into a genetic predictor has the ability to separate people out more across a PRS-distribution for a disease in terms of their risk of disease relative to others in a population (see Khera et al.&nbsp;2018, Nature Genetics (available on ELE page)).</p>
<p>In this part of the workshop, you will</p>
<ul>
<li>explore what happens to predictive accuracy (ROC-AUC) when relaxing the P-value threshold for SNP inclusion into the predictor (Study 1)</li>
<li>compare frequency of cases in deciles of the PRS determine to have the highest ROC-AUC (Study 2)</li>
<li>compare changes in odds-ratios on disease risk between those in the upper tails of the PRS-distribution vs everyone else to quantify if and how much individuals at the upper tail of the PRS distribution are more at risk (Study 2).</li>
</ul>
<p>The hypothesis is that there could be variants that are associated with our outcomes of interest (coronary artery disease) but have not reaching levels of statistical significance required to be regarded as associated due to the power of previous studies. These might help improve the predictive power of the PRS.</p>
</section>
<section id="generating-several-prs-variables" class="level1" data-number="9">
<h1 data-number="9"><span class="header-section-number">9</span> Generating several PRS variables</h1>
<p>We are going to generate several genetic predictors based on association results obtained from an independent published GWAS of CAD (see <code>CAD_GWAS.txt</code> for data). The data has been filtered to remove any variants with association <span class="math inline">\(P&gt;0.05\)</span>. The data has also been LD-pruned - i.e.&nbsp;genetic variants that are highly correlated with each as they are trasnmitted together through generations on haplotypes have been filtered whereby the most significant variant at a locus remains in the data.</p>
<p>We will load in the published GWAS data in <code>CAD_GWAS.txt</code>into a data frame called <code>cad_gwas_results</code>.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>cad_gwas_results <span class="ot">=</span> <span class="fu">read.table</span>(<span class="st">"CAD_GWAS.txt"</span>, <span class="at">header=</span>T, <span class="at">stringsAsFactors =</span> F)</span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a><span class="co"># preview</span></span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(cad_gwas_results)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>  MARKERNAME CHR     POS EFFECT_ALLELE OTHER_ALLELE EFFECT_ALLELE_FREQ     BETA        SE          P
1 rs13303118   1  918384             T            G           0.538848 0.025888 0.0125704 0.03945210
2     rs1921   1  949608             G            A           0.378821 0.040148 0.0133895 0.00271330
3 rs10910028   1 2147874             G            A           0.921892 0.050347 0.0188132 0.00744740
4   rs263533   1 2163568             C            T           0.568352 0.033695 0.0095875 0.00044060
5 rs72643476   1 2232532             G            A           0.837474 0.053983 0.0137838 0.00008990
6  rs2843152   1 2245570             C            G           0.686548 0.051921 0.0113505 0.00000478</code></pre>
</div>
</div>
<p>Let’s first check all the values in the <code>BETA</code> column are positive. If the estimates are all positive, this make using the effect estimate for the weight easier as the <code>EFFECT_ALLELE</code> reported in the GWAS results file will represent the risk-increasing allele.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="fu">min</span>(cad_gwas_results<span class="sc">$</span>BETA)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.021757</code></pre>
</div>
</div>
<p>As the smallest value is &gt;0, it means we do not need to alter the GWAS results data. If this was not the case, we would need to make the effect estimates postive and ensure the alleles associated with increasing disease risk are all stored in the same column - e.g.&nbsp;<code>EFFECT_ALLELE</code>.</p>
<p>Next, we will load a new CAD status phenotype file from a study called <code>study1</code>. As before, 1=case, 0=control.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>study1_pheno <span class="ot">=</span> <span class="fu">read.table</span>(<span class="st">"study1_cad_status.txt"</span>, <span class="at">header=</span>T, <span class="at">stringsAsFactors =</span> F)</span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a><span class="co"># preview</span></span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(study1_pheno)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>  id cad_status
1 P1          0
2 P2          1
3 P3          0
4 P4          1
5 P5          0
6 P6          0</code></pre>
</div>
</div>
<p>Now we will load in the genotype data available for individuals in <code>Study1</code> available in <code>study1_snp_genotypes.txt</code> into a DF called <code>study1_snp_genotypes</code>:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>study1_snp_genotypes <span class="ot">=</span> <span class="fu">read.table</span>(<span class="st">"study1_snp_genotypes.txt"</span>, <span class="at">header=</span>T, <span class="at">stringsAsFactors =</span> F)</span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a><span class="co"># preview</span></span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a>study1_snp_genotypes[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>      Variant Allele1 Allele2 P8770 P7908 P3280 P7171 P760 P421 P6175
1      rs1921       G       A     0     1     0     0    1    0     0
2  rs10910028       A       G     1     0     0     0    0    0     0
3    rs263533       C       T     0     1     1     1    2    1     2
4  rs72643476       G       A     1     0     0     0    1    1     0
5   rs2843152       C       G     2     1     1     1    2    1     2
6   rs4648559       C       T     2     2     2     1    2    0     1
7  rs75958134       G       A     0     0     0     0    0    0     1
8  rs35397508       C       T     1     2     1     2    2    1     1
9  rs61759167       C       T     0     0     0     0    0    1     2
10  rs1553287       G       A     0     0     0     1    0    0     0</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Get dimensions</span></span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(study1_snp_genotypes)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1]  7114 10003</code></pre>
</div>
</div>
<p>Next, merge the CAD GWAS association statistics onto the genotypes. We will be using the effect-estimates from the GWAS to weight the genotypes (similar to before).</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>study1_data_merged <span class="ot">=</span> <span class="fu">merge</span>(cad_gwas_results, study1_snp_genotypes, <span class="at">by.x =</span> <span class="st">"MARKERNAME"</span>, <span class="at">by.y=</span><span class="st">"Variant"</span>)</span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a><span class="co"># preview</span></span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a>study1_data_merged[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>]</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>   MARKERNAME CHR       POS EFFECT_ALLELE OTHER_ALLELE EFFECT_ALLELE_FREQ     BETA        SE         P Allele1 Allele2
1  rs10000834   4  11583158             G            A           0.801696 0.031356 0.0121174 0.0096626       A       G
2  rs10001629   4   7161941             G            C           0.976654 0.106786 0.0400934 0.0077349       G       C
3  rs10002176   4 125367078             G            A           0.706207 0.031781 0.0103446 0.0021247       G       A
4   rs1000423  17  59475642             T            C           0.667440 0.040272 0.0105814 0.0001413       C       T
5  rs10007118   4  38382277             A            G           0.932996 0.050856 0.0194857 0.0090563       G       A
6  rs10007560   4 184232222             T            C           0.812501 0.031550 0.0120838 0.0090293       T       C
7  rs10008867   4 144407046             A            G           0.627557 0.036359 0.0108487 0.0008038       G       A
8  rs10015472   4 113960961             G            A           0.638296 0.025041 0.0095245 0.0085607       G       A
9  rs10017826   4 156308870             T            C           0.979503 0.089639 0.0327882 0.0062593       T       C
10 rs10020278   4  17477882             G            A           0.958156 0.096245 0.0252571 0.0001386       G       A
   P8770 P7908 P3280 P7171 P760 P421 P6175 P5683 P3382
1      0     2     0     1    1    0     1     1     0
2      0     0     0     0    0    0     0     0     0
3      0     1     1     0    2    1     0     0     1
4      2     2     2     1    2    1     2     1     1
5      0     0     0     0    0    0     0     0     0
6      2     2     2     2    2    2     2     2     2
7      2     0     1     1    1    0     1     1     1
8      2     1     1     2    1    1     1     2     2
9      0     0     0     0    0    0     0     0     0
10     0     0     0     0    0    1     0     0     0</code></pre>
</div>
</div>
<p>Now we will weight the genotypes by the <code>BETA</code> column as this contains the weighting to apply to the genotypes. Like before, we must ensure the genotypes are coded to the allele the effect corresponds to in the GWAS dataset.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>study1_data_merged[study1_data_merged<span class="sc">$</span>EFFECT_ALLELE <span class="sc">!=</span> study1_data_merged<span class="sc">$</span>Allele2, <span class="dv">12</span><span class="sc">:</span><span class="fu">ncol</span>(study1_data_merged)] <span class="ot">=</span></span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a>   <span class="dv">2</span> <span class="sc">-</span> study1_data_merged[study1_data_merged<span class="sc">$</span>EFFECT_ALLELE <span class="sc">!=</span> study1_data_merged<span class="sc">$</span>Allele2, <span class="dv">12</span><span class="sc">:</span><span class="fu">ncol</span>(study1_data_merged)] </span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb46-4"><a href="#cb46-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Preview first 10 rows and columns:</span></span>
<span id="cb46-5"><a href="#cb46-5" aria-hidden="true" tabindex="-1"></a>study1_data_merged[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>]</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>   MARKERNAME CHR       POS EFFECT_ALLELE OTHER_ALLELE EFFECT_ALLELE_FREQ     BETA        SE         P Allele1 Allele2
1  rs10000834   4  11583158             G            A           0.801696 0.031356 0.0121174 0.0096626       A       G
2  rs10001629   4   7161941             G            C           0.976654 0.106786 0.0400934 0.0077349       G       C
3  rs10002176   4 125367078             G            A           0.706207 0.031781 0.0103446 0.0021247       G       A
4   rs1000423  17  59475642             T            C           0.667440 0.040272 0.0105814 0.0001413       C       T
5  rs10007118   4  38382277             A            G           0.932996 0.050856 0.0194857 0.0090563       G       A
6  rs10007560   4 184232222             T            C           0.812501 0.031550 0.0120838 0.0090293       T       C
7  rs10008867   4 144407046             A            G           0.627557 0.036359 0.0108487 0.0008038       G       A
8  rs10015472   4 113960961             G            A           0.638296 0.025041 0.0095245 0.0085607       G       A
9  rs10017826   4 156308870             T            C           0.979503 0.089639 0.0327882 0.0062593       T       C
10 rs10020278   4  17477882             G            A           0.958156 0.096245 0.0252571 0.0001386       G       A
   P8770 P7908 P3280 P7171 P760 P421 P6175 P5683 P3382
1      0     2     0     1    1    0     1     1     0
2      2     2     2     2    2    2     2     2     2
3      2     1     1     2    0    1     2     2     1
4      2     2     2     1    2    1     2     1     1
5      0     0     0     0    0    0     0     0     0
6      0     0     0     0    0    0     0     0     0
7      2     0     1     1    1    0     1     1     1
8      0     1     1     0    1    1     1     0     0
9      2     2     2     2    2    2     2     2     2
10     2     2     2     2    2    1     2     2     2</code></pre>
</div>
</div>
<p>Now we can create the weighted alleles based on multiplying the number of risk alleles by the effect of carrying each additional risk allele:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Copy DF containing genetic data related to CAD</span></span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a>study1_data_merged[<span class="dv">12</span><span class="sc">:</span><span class="fu">ncol</span>(study1_data_merged)] <span class="ot">=</span> study1_data_merged[,<span class="dv">12</span><span class="sc">:</span><span class="fu">ncol</span>(study1_data_merged)] <span class="sc">*</span> study1_data_merged<span class="sc">$</span>BETA</span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-4"><a href="#cb48-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Preview first 10 rows and columns:</span></span>
<span id="cb48-5"><a href="#cb48-5" aria-hidden="true" tabindex="-1"></a>study1_data_merged[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>]</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>   MARKERNAME CHR       POS EFFECT_ALLELE OTHER_ALLELE EFFECT_ALLELE_FREQ     BETA        SE         P Allele1 Allele2
1  rs10000834   4  11583158             G            A           0.801696 0.031356 0.0121174 0.0096626       A       G
2  rs10001629   4   7161941             G            C           0.976654 0.106786 0.0400934 0.0077349       G       C
3  rs10002176   4 125367078             G            A           0.706207 0.031781 0.0103446 0.0021247       G       A
4   rs1000423  17  59475642             T            C           0.667440 0.040272 0.0105814 0.0001413       C       T
5  rs10007118   4  38382277             A            G           0.932996 0.050856 0.0194857 0.0090563       G       A
6  rs10007560   4 184232222             T            C           0.812501 0.031550 0.0120838 0.0090293       T       C
7  rs10008867   4 144407046             A            G           0.627557 0.036359 0.0108487 0.0008038       G       A
8  rs10015472   4 113960961             G            A           0.638296 0.025041 0.0095245 0.0085607       G       A
9  rs10017826   4 156308870             T            C           0.979503 0.089639 0.0327882 0.0062593       T       C
10 rs10020278   4  17477882             G            A           0.958156 0.096245 0.0252571 0.0001386       G       A
      P8770    P7908    P3280    P7171     P760     P421    P6175    P5683    P3382
1  0.000000 0.062712 0.000000 0.031356 0.031356 0.000000 0.031356 0.031356 0.000000
2  0.213572 0.213572 0.213572 0.213572 0.213572 0.213572 0.213572 0.213572 0.213572
3  0.063562 0.031781 0.031781 0.063562 0.000000 0.031781 0.063562 0.063562 0.031781
4  0.080544 0.080544 0.080544 0.040272 0.080544 0.040272 0.080544 0.040272 0.040272
5  0.000000 0.000000 0.000000 0.000000 0.000000 0.000000 0.000000 0.000000 0.000000
6  0.000000 0.000000 0.000000 0.000000 0.000000 0.000000 0.000000 0.000000 0.000000
7  0.072718 0.000000 0.036359 0.036359 0.036359 0.000000 0.036359 0.036359 0.036359
8  0.000000 0.025041 0.025041 0.000000 0.025041 0.025041 0.025041 0.000000 0.000000
9  0.179278 0.179278 0.179278 0.179278 0.179278 0.179278 0.179278 0.179278 0.179278
10 0.192490 0.192490 0.192490 0.192490 0.192490 0.096245 0.192490 0.192490 0.192490</code></pre>
</div>
</div>
<p>Now we have our weighted alleles, we can sum different subsets of the variants based on the strength of their association in the CAD GWAS. First, we create a list of different P-value thresholds (a variant’s association P-value should be smaller than) to use for genetic variant inclusion to predictor. Let’s start with these thresholds:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a>ps <span class="ot">=</span> <span class="fu">c</span>(<span class="fl">5e-08</span>, <span class="fl">5e-07</span>, <span class="fl">5e-06</span>, <span class="fl">5e-05</span>, <span class="fl">5e-04</span>, <span class="fl">5e-03</span>, <span class="fl">0.05</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>Now we can cycle through the P-value threshold taking forward on genetic variants with an association P-threshold and calculating ROC-AUCs:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Get the next P-value threshold in the ps vector</span></span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (p <span class="cf">in</span> ps) {</span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb51-4"><a href="#cb51-4" aria-hidden="true" tabindex="-1"></a>  <span class="co"># subset the data so only variants with GWAS association P &lt; P-value threshold</span></span>
<span id="cb51-5"><a href="#cb51-5" aria-hidden="true" tabindex="-1"></a>  variant_subset <span class="ot">=</span> <span class="fu">subset</span>(study1_data_merged, P <span class="sc">&lt;</span> p)</span>
<span id="cb51-6"><a href="#cb51-6" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb51-7"><a href="#cb51-7" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Copy DF containing genetic data related to CAD</span></span>
<span id="cb51-8"><a href="#cb51-8" aria-hidden="true" tabindex="-1"></a>  cad_prs <span class="ot">=</span> <span class="fu">data.frame</span>(</span>
<span id="cb51-9"><a href="#cb51-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">id =</span> <span class="fu">names</span>(variant_subset)[<span class="dv">12</span><span class="sc">:</span><span class="fu">ncol</span>(variant_subset)],</span>
<span id="cb51-10"><a href="#cb51-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">prs =</span> <span class="fu">colSums</span>(variant_subset[<span class="dv">12</span><span class="sc">:</span><span class="fu">ncol</span>(variant_subset)]), </span>
<span id="cb51-11"><a href="#cb51-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">row.names =</span> <span class="cn">NULL</span></span>
<span id="cb51-12"><a href="#cb51-12" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb51-13"><a href="#cb51-13" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb51-14"><a href="#cb51-14" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Merege phenotype data into a DF</span></span>
<span id="cb51-15"><a href="#cb51-15" aria-hidden="true" tabindex="-1"></a>  cad_prs <span class="ot">=</span> <span class="fu">merge</span>(cad_prs, study1_pheno, <span class="at">by=</span><span class="st">"id"</span>)</span>
<span id="cb51-16"><a href="#cb51-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-17"><a href="#cb51-17" aria-hidden="true" tabindex="-1"></a>  <span class="co"># create roc object we can use for funtions associated with pROC</span></span>
<span id="cb51-18"><a href="#cb51-18" aria-hidden="true" tabindex="-1"></a>  roc_obj <span class="ot">=</span> <span class="fu">roc</span>(cad_prs<span class="sc">$</span>cad_status , cad_prs<span class="sc">$</span>prs, <span class="at">quiet =</span> T)</span>
<span id="cb51-19"><a href="#cb51-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-20"><a href="#cb51-20" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Calculate the area under the curve:</span></span>
<span id="cb51-21"><a href="#cb51-21" aria-hidden="true" tabindex="-1"></a>  auc_value <span class="ot">=</span> <span class="fu">auc</span>(roc_obj)</span>
<span id="cb51-22"><a href="#cb51-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="fu">paste0</span>(<span class="st">"Minimum P-value for SNPs to be included in PRS: "</span>, p))</span>
<span id="cb51-23"><a href="#cb51-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="fu">paste0</span>(<span class="st">"ROC-AUC: "</span>, <span class="fu">round</span>(auc_value, <span class="dv">4</span>)))</span>
<span id="cb51-24"><a href="#cb51-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-25"><a href="#cb51-25" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Minimum P-value for SNPs to be included in PRS: 5e-08"
[1] "ROC-AUC: 0.561"
[1] "Minimum P-value for SNPs to be included in PRS: 5e-07"
[1] "ROC-AUC: 0.5651"
[1] "Minimum P-value for SNPs to be included in PRS: 5e-06"
[1] "ROC-AUC: 0.572"
[1] "Minimum P-value for SNPs to be included in PRS: 5e-05"
[1] "ROC-AUC: 0.5765"
[1] "Minimum P-value for SNPs to be included in PRS: 5e-04"
[1] "ROC-AUC: 0.5801"
[1] "Minimum P-value for SNPs to be included in PRS: 0.005"
[1] "ROC-AUC: 0.5768"
[1] "Minimum P-value for SNPs to be included in PRS: 0.05"
[1] "ROC-AUC: 0.5767"</code></pre>
</div>
</div>
<section id="questions-1" class="level2" data-number="9.1">
<h2 data-number="9.1" class="anchored" data-anchor-id="questions-1"><span class="header-section-number">9.1</span> Questions:</h2>
<ul>
<li>Which P-value threshold has yielded a GRS with the highest prediction accuracy?</li>
<li>Do you see any pattern in the ROC-AUCs across different P-value thresholds? If so, what do you think might be causing it?</li>
</ul>
<p><strong>Return to the course page to post your thoughts</strong></p>
<hr>
</section>
</section>
<section id="taking-the-p-value-threshold-forwards-into-an-independent-study" class="level1" data-number="10">
<h1 data-number="10"><span class="header-section-number">10</span> Taking the P-value threshold forwards into an independent study</h1>
<p>We will use the P-value that returned the highest ROC-curve AUC to generate a GRS for CAD in an independent study and assess prevalence of disease across deciles of the GRS, odds ratios for disease risk in upper tails of the GRS distribution, and it’s effectiveness at discriminating between cases and controls.</p>
<p>Load the CAD status of individuals in <code>Study2</code>:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a>study2_pheno <span class="ot">=</span> <span class="fu">read.table</span>(<span class="st">"study2_cad_status.txt"</span>, <span class="at">header=</span>T, <span class="at">stringsAsFactors =</span> F)</span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a><span class="co"># preview</span></span>
<span id="cb53-3"><a href="#cb53-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(study2_pheno)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>  id cad_status
1 P1          0
2 P2          0
3 P3          0
4 P4          1
5 P5          0
6 P6          0</code></pre>
</div>
</div>
<p>Now load the genotypes for individuals in <code>Study2</code> stored in <code>study2_snp_genotypes.txt</code> into a DF called <code>study2_snp_genotypes</code>.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a>study2_snp_genotypes <span class="ot">=</span> <span class="fu">read.table</span>(<span class="st">"study2_snp_genotypes.txt"</span>, <span class="at">header=</span>T, <span class="at">stringsAsFactors =</span> F)</span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a><span class="co"># preview</span></span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a>study2_snp_genotypes[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>      Variant Allele1 Allele2 P7212 P3362 P4481 P1345 P4276 P89 P4066
1      rs1921       G       A     2     0     1     0     1   1     1
2  rs10910028       A       G     0     0     1     0     0   0     0
3    rs263533       C       T     2     2     1     2     0   2     0
4  rs72643476       G       A     0     1     1     2     0   0     0
5   rs2843152       C       G     2     2     2     2     0   1     0
6   rs4648559       C       T     1     2     2     1     1   2     2
7  rs75958134       G       A     0     0     0     0     0   0     0
8  rs35397508       C       T     1     0     0     0     1   0     1
9  rs61759167       C       T     1     0     1     0     1   0     0
10  rs1553287       G       A     0     0     0     1     1   0     0</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Get dimensions</span></span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(study2_snp_genotypes)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 7114 9769</code></pre>
</div>
</div>
<p>Then merge the GWAS results containing the allele-effect estimates to the genotypes on CAD risk.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a>study2_data_merged <span class="ot">=</span> <span class="fu">merge</span>(cad_gwas_results, study2_snp_genotypes, </span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a>                           <span class="at">by.x=</span><span class="st">"MARKERNAME"</span>, <span class="at">by.y=</span><span class="st">"Variant"</span>)</span>
<span id="cb59-3"><a href="#cb59-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-4"><a href="#cb59-4" aria-hidden="true" tabindex="-1"></a>study2_data_merged[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>]</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>   MARKERNAME CHR       POS EFFECT_ALLELE OTHER_ALLELE EFFECT_ALLELE_FREQ     BETA        SE         P Allele1 Allele2
1  rs10000834   4  11583158             G            A           0.801696 0.031356 0.0121174 0.0096626       A       G
2  rs10001629   4   7161941             G            C           0.976654 0.106786 0.0400934 0.0077349       G       C
3  rs10002176   4 125367078             G            A           0.706207 0.031781 0.0103446 0.0021247       G       A
4   rs1000423  17  59475642             T            C           0.667440 0.040272 0.0105814 0.0001413       C       T
5  rs10007118   4  38382277             A            G           0.932996 0.050856 0.0194857 0.0090563       G       A
6  rs10007560   4 184232222             T            C           0.812501 0.031550 0.0120838 0.0090293       T       C
7  rs10008867   4 144407046             A            G           0.627557 0.036359 0.0108487 0.0008038       G       A
8  rs10015472   4 113960961             G            A           0.638296 0.025041 0.0095245 0.0085607       G       A
9  rs10017826   4 156308870             T            C           0.979503 0.089639 0.0327882 0.0062593       T       C
10 rs10020278   4  17477882             G            A           0.958156 0.096245 0.0252571 0.0001386       G       A
   P7212 P3362 P4481 P1345 P4276 P89 P4066 P1160 P6249
1      1     0     0     1     0   0     0     0     1
2      0     0     0     0     0   0     0     0     1
3      1     0     0     0     1   0     1     1     0
4      2     1     2     1     1   2     1     2     1
5      1     0     0     1     0   0     0     0     0
6      1     1     2     2     2   2     0     2     2
7      1     0     1     0     0   2     1     0     1
8      2     2     1     1     2   2     1     1     2
9      0     0     0     0     0   0     0     0     0
10     0     0     0     0     0   0     0     0     0</code></pre>
</div>
</div>
<p>Now we will weight the genotypes by the <code>BETA</code> column as this contains the weighting to apply to the genotypes. Like before, we must ensure the genotypes are coded to the allele the effect corresponds to in the GWAS dataset.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a>study2_data_merged[study2_data_merged<span class="sc">$</span>EFFECT_ALLELE <span class="sc">!=</span> study2_data_merged<span class="sc">$</span>Allele2, <span class="dv">12</span><span class="sc">:</span><span class="fu">ncol</span>(study2_data_merged)] <span class="ot">=</span></span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a>   <span class="dv">2</span> <span class="sc">-</span> study2_data_merged[study2_data_merged<span class="sc">$</span>EFFECT_ALLELE <span class="sc">!=</span> study2_data_merged<span class="sc">$</span>Allele2, <span class="dv">12</span><span class="sc">:</span><span class="fu">ncol</span>(study2_data_merged)] </span>
<span id="cb61-3"><a href="#cb61-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-4"><a href="#cb61-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Preview first 10 rows and columns:</span></span>
<span id="cb61-5"><a href="#cb61-5" aria-hidden="true" tabindex="-1"></a>study2_data_merged[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>]</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>   MARKERNAME CHR       POS EFFECT_ALLELE OTHER_ALLELE EFFECT_ALLELE_FREQ     BETA        SE         P Allele1 Allele2
1  rs10000834   4  11583158             G            A           0.801696 0.031356 0.0121174 0.0096626       A       G
2  rs10001629   4   7161941             G            C           0.976654 0.106786 0.0400934 0.0077349       G       C
3  rs10002176   4 125367078             G            A           0.706207 0.031781 0.0103446 0.0021247       G       A
4   rs1000423  17  59475642             T            C           0.667440 0.040272 0.0105814 0.0001413       C       T
5  rs10007118   4  38382277             A            G           0.932996 0.050856 0.0194857 0.0090563       G       A
6  rs10007560   4 184232222             T            C           0.812501 0.031550 0.0120838 0.0090293       T       C
7  rs10008867   4 144407046             A            G           0.627557 0.036359 0.0108487 0.0008038       G       A
8  rs10015472   4 113960961             G            A           0.638296 0.025041 0.0095245 0.0085607       G       A
9  rs10017826   4 156308870             T            C           0.979503 0.089639 0.0327882 0.0062593       T       C
10 rs10020278   4  17477882             G            A           0.958156 0.096245 0.0252571 0.0001386       G       A
   P7212 P3362 P4481 P1345 P4276 P89 P4066 P1160 P6249
1      1     0     0     1     0   0     0     0     1
2      2     2     2     2     2   2     2     2     1
3      1     2     2     2     1   2     1     1     2
4      2     1     2     1     1   2     1     2     1
5      1     0     0     1     0   0     0     0     0
6      1     1     0     0     0   0     2     0     0
7      1     0     1     0     0   2     1     0     1
8      0     0     1     1     0   0     1     1     0
9      2     2     2     2     2   2     2     2     2
10     2     2     2     2     2   2     2     2     2</code></pre>
</div>
</div>
<p>Now we are ready to create the GRS in <code>Study2</code> using SNPs passing the P-value threshold that resulted in a predictor in <code>Study1</code> that had the highest AUC. For example, if the P-value threshold associated with the highest ROC-curve AUC was <span class="math inline">\(P&lt;5e-04\)</span>, then we would use something like below:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a><span class="co"># subset the data so only variants with GWAS association P &lt; P-value threshold</span></span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a>variant_subset <span class="ot">=</span> <span class="fu">subset</span>(study2_data_merged, P <span class="sc">&lt;</span> <span class="fl">5e-04</span>)</span>
<span id="cb63-3"><a href="#cb63-3" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb63-4"><a href="#cb63-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Copy DF containing genetic data related to CAD</span></span>
<span id="cb63-5"><a href="#cb63-5" aria-hidden="true" tabindex="-1"></a>cad_prs <span class="ot">=</span> <span class="fu">data.frame</span>(</span>
<span id="cb63-6"><a href="#cb63-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">id =</span> <span class="fu">names</span>(variant_subset)[<span class="dv">12</span><span class="sc">:</span><span class="fu">ncol</span>(variant_subset)],</span>
<span id="cb63-7"><a href="#cb63-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">prs =</span> <span class="fu">colSums</span>(variant_subset[<span class="dv">12</span><span class="sc">:</span><span class="fu">ncol</span>(variant_subset)]), </span>
<span id="cb63-8"><a href="#cb63-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">row.names =</span> <span class="cn">NULL</span></span>
<span id="cb63-9"><a href="#cb63-9" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb63-10"><a href="#cb63-10" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb63-11"><a href="#cb63-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Merge phenotype data into a DF</span></span>
<span id="cb63-12"><a href="#cb63-12" aria-hidden="true" tabindex="-1"></a>cad_prs <span class="ot">=</span> <span class="fu">merge</span>(cad_prs, study2_pheno, <span class="at">by=</span><span class="st">"id"</span>)</span>
<span id="cb63-13"><a href="#cb63-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-14"><a href="#cb63-14" aria-hidden="true" tabindex="-1"></a><span class="co"># create roc object we can use for funtions associated with pROC</span></span>
<span id="cb63-15"><a href="#cb63-15" aria-hidden="true" tabindex="-1"></a>roc_obj <span class="ot">=</span> <span class="fu">roc</span>(cad_prs<span class="sc">$</span>cad_status , cad_prs<span class="sc">$</span>prs, <span class="at">quiet =</span> T)</span>
<span id="cb63-16"><a href="#cb63-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-17"><a href="#cb63-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate the area under the curve:</span></span>
<span id="cb63-18"><a href="#cb63-18" aria-hidden="true" tabindex="-1"></a>auc_value <span class="ot">=</span> <span class="fu">auc</span>(roc_obj)</span>
<span id="cb63-19"><a href="#cb63-19" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste0</span>(<span class="st">"Minimum P-value for SNPs to be included in PRS: "</span>, <span class="fl">5e-04</span>))</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Minimum P-value for SNPs to be included in PRS: 5e-04"</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste0</span>(<span class="st">"ROC-AUC: "</span>, <span class="fu">round</span>(auc_value, <span class="dv">4</span>)))</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "ROC-AUC: 0.5738"</code></pre>
</div>
</div>
<p>We can combine the GRS with CAD status and plot the GRS distribution between cases and controls similar to before to get a quick visualisation on how separated cases and controls are on the GRS distribution:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(</span>
<span id="cb67-2"><a href="#cb67-2" aria-hidden="true" tabindex="-1"></a>  cad_prs, <span class="fu">aes</span>(<span class="at">x=</span>prs, <span class="at">fill =</span> <span class="fu">as.factor</span>(cad_status))) <span class="sc">+</span> </span>
<span id="cb67-3"><a href="#cb67-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">position =</span> <span class="st">"identity"</span>, <span class="at">alpha =</span> <span class="fl">0.5</span>, <span class="at">bins =</span> <span class="dv">30</span>) <span class="sc">+</span> </span>
<span id="cb67-4"><a href="#cb67-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb67-5"><a href="#cb67-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(</span>
<span id="cb67-6"><a href="#cb67-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> <span class="st">"CAD PRS"</span>,</span>
<span id="cb67-7"><a href="#cb67-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> <span class="st">"N individuals"</span>,</span>
<span id="cb67-8"><a href="#cb67-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">fill =</span> <span class="st">"Case / control status"</span></span>
<span id="cb67-9"><a href="#cb67-9" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="hpdm208z_workshop9_files/figure-html/unnamed-chunk-32-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<section id="questions-and-tasks" class="level2" data-number="10.1">
<h2 data-number="10.1" class="anchored" data-anchor-id="questions-and-tasks"><span class="header-section-number">10.1</span> Questions and tasks</h2>
<ul>
<li>How does the AUC in <code>Study2</code> compare to what you observed in <code>Study1</code>?</li>
<li>How does the AUC in <code>Study2</code> compare to what you observed based on the first PRS you created for CAD?</li>
<li>Draw the ROC-curve for the PRS geerated in <code>study2</code></li>
</ul>
<p><strong>Return to the course page to post your thoughts and screenshot of the ROC curve</strong></p>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
      const outerScaffold = trigger.parentElement.cloneNode(true);
      const codeEl = outerScaffold.querySelector('code');
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




</body></html>