<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.8.25">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>HPDM208Z Stratified Medicine Workshop 7: Introduction to the analysis of genetic data</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="hpdm208z_workshop7_files/libs/clipboard/clipboard.min.js"></script>
<script src="hpdm208z_workshop7_files/libs/quarto-html/quarto.js" type="module"></script>
<script src="hpdm208z_workshop7_files/libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="hpdm208z_workshop7_files/libs/quarto-html/axe/axe-check.js" type="module"></script>
<script src="hpdm208z_workshop7_files/libs/quarto-html/popper.min.js"></script>
<script src="hpdm208z_workshop7_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="hpdm208z_workshop7_files/libs/quarto-html/anchor.min.js"></script>
<link href="hpdm208z_workshop7_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="hpdm208z_workshop7_files/libs/quarto-html/quarto-syntax-highlighting-7b89279ff1a6dce999919e0e67d4d9ec.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="hpdm208z_workshop7_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="hpdm208z_workshop7_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="hpdm208z_workshop7_files/libs/bootstrap/bootstrap-d5214d16172734abcd1dcccc621ce81b.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">


</head>

<body class="quarto-light">

<header id="title-block-header" class="quarto-title-block default page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <h1 class="title">HPDM208Z Stratified Medicine Workshop 7: Introduction to the analysis of genetic data</h1>
            <p class="subtitle lead">MSc Health Data Science, University of Exeter</p>
                      </div>
  </div>
    
  
  <div class="quarto-title-meta">

      
    
      
    </div>
    
  
  </header><div id="quarto-content" class="page-columns page-rows-contents page-layout-article">
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
  <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction"><span class="header-section-number">1</span> Introduction</a></li>
  <li><a href="#aims-of-workshop" id="toc-aims-of-workshop" class="nav-link" data-scroll-target="#aims-of-workshop"><span class="header-section-number">2</span> Aims of workshop</a></li>
  <li><a href="#conda-environment" id="toc-conda-environment" class="nav-link" data-scroll-target="#conda-environment"><span class="header-section-number">3</span> Conda environment</a></li>
  <li><a href="#workshop-directory" id="toc-workshop-directory" class="nav-link" data-scroll-target="#workshop-directory"><span class="header-section-number">4</span> Workshop directory</a></li>
  <li><a href="#workshop-data" id="toc-workshop-data" class="nav-link" data-scroll-target="#workshop-data"><span class="header-section-number">5</span> Workshop data</a></li>
  <li><a href="#calculating-allele-and-genotype-frequencies" id="toc-calculating-allele-and-genotype-frequencies" class="nav-link" data-scroll-target="#calculating-allele-and-genotype-frequencies"><span class="header-section-number">6</span> Calculating allele and genotype frequencies</a>
  <ul class="collapse">
  <li><a href="#reading-genotypes-using-pandas" id="toc-reading-genotypes-using-pandas" class="nav-link" data-scroll-target="#reading-genotypes-using-pandas"><span class="header-section-number">6.1</span> Reading genotypes using Pandas</a></li>
  <li><a href="#converting-genotypes-to-numerical-representation" id="toc-converting-genotypes-to-numerical-representation" class="nav-link" data-scroll-target="#converting-genotypes-to-numerical-representation"><span class="header-section-number">6.2</span> Converting genotypes to numerical representation</a></li>
  <li><a href="#tasks" id="toc-tasks" class="nav-link" data-scroll-target="#tasks"><span class="header-section-number">6.3</span> Tasks</a></li>
  </ul></li>
  <li><a href="#using-plink-to-analyse-genotype-data" id="toc-using-plink-to-analyse-genotype-data" class="nav-link" data-scroll-target="#using-plink-to-analyse-genotype-data"><span class="header-section-number">7</span> Using PLINK to analyse genotype data</a>
  <ul class="collapse">
  <li><a href="#file-descriptors" id="toc-file-descriptors" class="nav-link" data-scroll-target="#file-descriptors"><span class="header-section-number">7.1</span> File descriptors</a>
  <ul class="collapse">
  <li><a href="#fam-file" id="toc-fam-file" class="nav-link" data-scroll-target="#fam-file"><span class="header-section-number">7.1.1</span> .fam file</a></li>
  <li><a href="#bim-file" id="toc-bim-file" class="nav-link" data-scroll-target="#bim-file"><span class="header-section-number">7.1.2</span> .bim file</a></li>
  <li><a href="#bed-file" id="toc-bed-file" class="nav-link" data-scroll-target="#bed-file"><span class="header-section-number">7.1.3</span> .bed file</a></li>
  </ul></li>
  <li><a href="#use-plink-to-read-the-files-and-calculate-allele-frequencies" id="toc-use-plink-to-read-the-files-and-calculate-allele-frequencies" class="nav-link" data-scroll-target="#use-plink-to-read-the-files-and-calculate-allele-frequencies"><span class="header-section-number">7.2</span> Use PLINK to read the files and calculate allele frequencies</a>
  <ul class="collapse">
  <li><a href="#task" id="toc-task" class="nav-link" data-scroll-target="#task"><span class="header-section-number">7.2.1</span> Task</a></li>
  </ul></li>
  <li><a href="#use-plink-to-convert-the-non-human-readble-genotype-data-to-human-readable-format-for-analysis" id="toc-use-plink-to-convert-the-non-human-readble-genotype-data-to-human-readable-format-for-analysis" class="nav-link" data-scroll-target="#use-plink-to-convert-the-non-human-readble-genotype-data-to-human-readable-format-for-analysis"><span class="header-section-number">7.3</span> Use PLINK to convert the non-human readble genotype data to human-readable format for analysis</a></li>
  <li><a href="#subsetting-variants-and-individuals" id="toc-subsetting-variants-and-individuals" class="nav-link" data-scroll-target="#subsetting-variants-and-individuals"><span class="header-section-number">7.4</span> Subsetting variants and individuals</a>
  <ul class="collapse">
  <li><a href="#task-1" id="toc-task-1" class="nav-link" data-scroll-target="#task-1"><span class="header-section-number">7.4.1</span> Task</a></li>
  </ul></li>
  <li><a href="#changing-the-genotype-codes-to-represent-numbers-of-a-specific-allele" id="toc-changing-the-genotype-codes-to-represent-numbers-of-a-specific-allele" class="nav-link" data-scroll-target="#changing-the-genotype-codes-to-represent-numbers-of-a-specific-allele"><span class="header-section-number">7.5</span> Changing the genotype codes to represent numbers of a specific allele</a></li>
  <li><a href="#task-2" id="toc-task-2" class="nav-link" data-scroll-target="#task-2"><span class="header-section-number">7.6</span> Task</a></li>
  <li><a href="#running-a-genetic-association-analysis" id="toc-running-a-genetic-association-analysis" class="nav-link" data-scroll-target="#running-a-genetic-association-analysis"><span class="header-section-number">7.7</span> Running a genetic association analysis</a>
  <ul class="collapse">
  <li><a href="#questions" id="toc-questions" class="nav-link" data-scroll-target="#questions"><span class="header-section-number">7.7.1</span> Questions</a></li>
  </ul></li>
  </ul></li>
  </ul>
</nav>
</div>
<main class="content quarto-banner-title-block" id="quarto-document-content">





<section id="introduction" class="level1" data-number="1">
<h1 data-number="1"><span class="header-section-number">1</span> Introduction</h1>
<p>In this workshop, you will handle data associated with genetic variation to reinfoce some of the primary concepts taught this week.</p>
</section>
<section id="aims-of-workshop" class="level1" data-number="2">
<h1 data-number="2"><span class="header-section-number">2</span> Aims of workshop</h1>
<ul>
<li>reinforce the concepts of alleles, genotypes, and their frequencies</li>
<li>teach you how to perform basic analysis of genotype data in Python</li>
<li>introduce you to PLINK - a commonly used tool for processing genetic data</li>
</ul>
</section>
<section id="conda-environment" class="level1" data-number="3">
<h1 data-number="3"><span class="header-section-number">3</span> Conda environment</h1>
<p>Before attempting this workshop, you should activate the <code>hpdm208z</code> conda environment that will provide the R through Jupyter Lab and required library dependencies:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate hpdm208z</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="workshop-directory" class="level1" data-number="4">
<h1 data-number="4"><span class="header-section-number">4</span> Workshop directory</h1>
<p>All relevant files associated with this workshops can be found in the following directory:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb2"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="ex">/home/ubuntu/hpdm208z/workshops/genetic_data/</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>You should navigate to this directory prior to starting the workshop. Note, a jupyter notebook can be be loaded from this directory to assist you in undertaking this workshop:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb3"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="ex">genetic_data.ipynb</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="workshop-data" class="level1" data-number="5">
<h1 data-number="5"><span class="header-section-number">5</span> Workshop data</h1>
<p>Below is a table describing the data files in the workshops directory:</p>
<table class="caption-top table">
<colgroup>
<col style="width: 30%">
<col style="width: 70%">
</colgroup>
<thead>
<tr class="header">
<th>File Name</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>snps.txt</code></td>
<td>Contains information on 5 single nucleotide polymorphisms</td>
</tr>
<tr class="even">
<td><code>genotypes.txt</code></td>
<td>Contains genotypes for the 5 SNPs listed in <code>snps.txt</code> for 10 individuals</td>
</tr>
<tr class="odd">
<td><code>workshop.[bed|bim|fam]</code></td>
<td>A file set containing genotype data in binary format for 5 genetics variants in 10 individuals</td>
</tr>
<tr class="even">
<td><code>t2d_study_5k.[bed|bim|fam]</code></td>
<td>A file set containing genotype data in binary format for 5000 individuals and 10,000 genetic variants</td>
</tr>
</tbody>
</table>
<hr>
</section>
<section id="calculating-allele-and-genotype-frequencies" class="level1" data-number="6">
<h1 data-number="6"><span class="header-section-number">6</span> Calculating allele and genotype frequencies</h1>
<section id="reading-genotypes-using-pandas" class="level2" data-number="6.1">
<h2 data-number="6.1" class="anchored" data-anchor-id="reading-genotypes-using-pandas"><span class="header-section-number">6.1</span> Reading genotypes using Pandas</h2>
<p>First, import the data in <code>snps.txt</code> file into an pandas data frame called <code>snps</code>.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb4"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Import pandas library</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Read data into a pandas df called snps</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>snps <span class="op">=</span> pd.read_csv(<span class="st">"snps.txt"</span>, sep<span class="op">=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, index_col<span class="op">=</span><span class="st">"rsID"</span>)</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a><span class="co"># preview</span></span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(snps.to_string(index<span class="op">=</span><span class="va">False</span>))</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code> chromosome  base_pair_position_b37 nearest_gene allele_1 allele_2
          1               177889480       SEC16B        G        A
          2                  632348       TMEM18        A        G
          4                45182527       GNPDA2        G        A
         16                53803574          FTO        A        T
         18                57829135         MC4R        C        T</code></pre>
</div>
</div>
<p>The fields in the dataframe are described below:</p>
<table class="caption-top table">
<colgroup>
<col style="width: 30%">
<col style="width: 70%">
</colgroup>
<thead>
<tr class="header">
<th>Field</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>rsID</code></td>
<td>Unique identifier of genetic variant</td>
</tr>
<tr class="even">
<td><code>chromosome</code></td>
<td>Chromosome number the genetic variant is located</td>
</tr>
<tr class="odd">
<td><code>base_pair_position_b37</code></td>
<td>Base-pair position along the chromosome the genetic variant is located (mapped to build-37 of human genome)</td>
</tr>
<tr class="even">
<td><code>nearest_gene</code></td>
<td>Name of the nearest gene (i.e.&nbsp;protein coding sequence) to the genetic variant</td>
</tr>
<tr class="odd">
<td><code>allele_1</code></td>
<td>One of the alleles observed at the SNP site</td>
</tr>
<tr class="even">
<td><code>allele_2</code></td>
<td>The other allele observed at the SNP site</td>
</tr>
</tbody>
</table>
<p>Note: the nucleotides associated with SNP and listed in the <code>allele_1</code> and <code>allele_2</code> columns have been randomly assigned. We could easily swap the nucleotides listed in these two columns for each respective genetic variant.</p>
<p>Next, import the data in the <code>genotypes.txt</code> file into a data frame called <code>genotypes</code>. This file contains data on the genotypes for 10 individuals for the 5 genetic variants we have data on in the <code>snps</code> data frame above. Remember, genotypes are simply allele combinations carried by each individual at the site of a genetic variant. At each variant site, some individuals will carry one of the following:</p>
<ul>
<li>two copies of <code>allele_1</code> (these individuals are homozygous for <code>allele_1</code>)</li>
<li>one copy of <code>allele_1</code> and one copy of <code>allele_2</code> (these individuals are heterozygous)</li>
<li>two copies of <code>allele_2</code> (these individuals are homozygous for allele_2a)</li>
</ul>
<p>Import data in <code>genotypes.txt</code> into a data frame called <code>genotypes</code></p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>genotypes <span class="op">=</span> pd.read_csv(<span class="st">"genotypes.txt"</span>, sep<span class="op">=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>,  index_col<span class="op">=</span><span class="st">"Ind_ID"</span>)</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="co"># preview</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(genotypes.to_string(index<span class="op">=</span><span class="va">False</span>))</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>rs543874 rs13021737 rs10938397 rs1558902 rs6567160
      GG         AA         GA        AT        TT
      GA         GG         AA        TT        TC
      AA         AG         GG        AT        CC
      GG         GG         AA        AT        TC
      AA         AG         AA        TT        TC
      AA         AG         AA        AT        CC
      AG         GG         AA        AT        TC
      AA         GG         AA        AT        TT
      GG         GG         GA        TT        TT
      AA         AG         AA        TT        TT</code></pre>
</div>
</div>
</section>
<section id="converting-genotypes-to-numerical-representation" class="level2" data-number="6.2">
<h2 data-number="6.2" class="anchored" data-anchor-id="converting-genotypes-to-numerical-representation"><span class="header-section-number">6.2</span> Converting genotypes to numerical representation</h2>
<p>Presently, the genotypes (i.e.&nbsp;combination of alleles carried) are coded as nucleotide bases (abbreviated) at the SNP positions in the genome. As we are going to perform some calculations using this data, we will convert these genotype codes to a numerical representation of the number of alleles carried that are labelled <code>allele_1</code> in the <code>snps</code> data frame.</p>
<p>To do this will create another DataFrame called <code>genotypes_numeric</code> that will store the data we will be able to analyze directly. As human are diploid, genotypes on an person-level can only take 1 of 3 values: <code>0</code>, <code>1</code> or <code>2</code>.</p>
<p>The following code above looks up the number of occurrences of the character held in <code>allele_1</code> of the <code>snps</code> data frame for each genotype in the <code>genotypes_numeric</code> dataframe for the respective genetic variant. It then returns the count as the genotype coding (i.e.&nbsp;the number of <code>allele_1</code>’s carried for each individual for each genetic variant).</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Copy the data frame</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>genotypes_numeric <span class="op">=</span> pd.DataFrame(genotypes)</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(genotypes_numeric)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>       rs543874 rs13021737 rs10938397 rs1558902 rs6567160
Ind_ID                                                   
A            GG         AA         GA        AT        TT
B            GA         GG         AA        TT        TC
C            AA         AG         GG        AT        CC
D            GG         GG         AA        AT        TC
E            AA         AG         AA        TT        TC
F            AA         AG         AA        AT        CC
G            AG         GG         AA        AT        TC
H            AA         GG         AA        AT        TT
I            GG         GG         GA        TT        TT
J            AA         AG         AA        TT        TT</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> snp <span class="kw">in</span> genotypes_numeric:</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>    genotypes_numeric[snp] <span class="op">=</span> [ x.count(snps.loc[snp, <span class="st">"allele_1"</span>]) <span class="cf">for</span> x <span class="kw">in</span> genotypes_numeric[ snp ] ]</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(genotypes_numeric) </span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>        rs543874  rs13021737  rs10938397  rs1558902  rs6567160
Ind_ID                                                        
A              2           2           1          1          0
B              1           0           0          0          1
C              0           1           2          1          2
D              2           0           0          1          1
E              0           1           0          0          1
F              0           1           0          1          2
G              1           0           0          1          1
H              0           0           0          1          0
I              2           0           1          0          0
J              0           1           0          0          0</code></pre>
</div>
</div>
</section>
<section id="tasks" class="level2" data-number="6.3">
<h2 data-number="6.3" class="anchored" data-anchor-id="tasks"><span class="header-section-number">6.3</span> Tasks</h2>
<ul>
<li><p>Task 1. Write Python statements to calculate the allele frequencies for <code>allele_1</code> and <code>allele_2</code> for each variant using the <code>genotypes_numeric</code> data frame. Add the frequencies calculated to the <code>snps</code> data frame. Also add <code>minor_allele</code> and <code>major_allele</code> columns to the <code>snps</code> data frame based on the allele frequency calculations.</p></li>
<li><p>Task 2. Write python statements to calculate the genotype frequencies of each genetic variant using the <code>genotypes_numeric</code> data frame You should insert the genotype frequencies calculated into the <code>snps</code> data frame.</p></li>
</ul>
<p>If would you would prefer, you can calculate by hand and enter your answers into the <code>Genotype_Data_Workshop_Tasks.docx</code> document available on the ELE page. The main aim of this task is to help you understand the difference between alleles and genotypes.</p>
</section>
</section>
<section id="using-plink-to-analyse-genotype-data" class="level1" data-number="7">
<h1 data-number="7"><span class="header-section-number">7</span> Using PLINK to analyse genotype data</h1>
<p>A widely used tool in the genetics research community for wrangling, cleaning and summarizing genetic data is <a href="https://www.cog-genomics.org/plink/">PLINK</a>. PLINK has been installed on your virtual machines can be executed using the following command:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb12"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="ex">/home/ubuntu/hpdm208z/programs/plink</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>We are going to use PLINK to read in the <code>workshop.[bed|bim|fam]</code> which contain the information regarding the 5 SNPs above and genotype information for these SNPs on the same 10 individuals. These files represent one of the most common used ways of representing genotype data. The table below describes each of these files.</p>
<table class="caption-top table">
<colgroup>
<col style="width: 22%">
<col style="width: 77%">
</colgroup>
<thead>
<tr class="header">
<th>File</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>workshop.fam</code></td>
<td>Contains information about the individuals with genotype data</td>
</tr>
<tr class="even">
<td><code>workshop.bim</code></td>
<td>Contains information about the genetic variants the genotype data refers to</td>
</tr>
<tr class="odd">
<td><code>workshop.bed</code></td>
<td>Contains non-human readble (bit-pair) representation of genotypes for variants listed in the <code>.bim</code> file for individuals in the <code>.fam</code> file. Ordering of genotypes are defined by the the order found in the <code>.bed</code> and <code>.fam</code> file</td>
</tr>
</tbody>
</table>
<section id="file-descriptors" class="level2" data-number="7.1">
<h2 data-number="7.1" class="anchored" data-anchor-id="file-descriptors"><span class="header-section-number">7.1</span> File descriptors</h2>
<p>A full set of file descriptors accepted by PLINK can be found <a href="https://www.cog-genomics.org/plink/2.0/formats">here</a> and are briefy described below.</p>
<section id="fam-file" class="level3" data-number="7.1.1">
<h3 data-number="7.1.1" class="anchored" data-anchor-id="fam-file"><span class="header-section-number">7.1.1</span> .fam file</h3>
<p>The <code>.fam</code> file contains information the individuals we have genotype data for. The following table describes the columns in the <code>.fam</code> file that do not have a header</p>
<table class="caption-top table">
<colgroup>
<col style="width: 13%">
<col style="width: 86%">
</colgroup>
<thead>
<tr class="header">
<th>Column #</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>1</td>
<td>Family ID (FID) - value shared between individuals to indicate some form of relation exists. Unique to individual if not related</td>
</tr>
<tr class="even">
<td>2</td>
<td>Individual ID (IID) - Unique identifier specific to individual</td>
</tr>
<tr class="odd">
<td>3</td>
<td>Paternal ID - Can be set to IID of father if in data when required - otherwise set to 0</td>
</tr>
<tr class="even">
<td>4</td>
<td>Maternal ID - Can be set to IID of mother if in data when required - otherwise set to 0</td>
</tr>
<tr class="odd">
<td>5</td>
<td>Sex - Reported sex of individual - 0 if missing</td>
</tr>
<tr class="even">
<td>6</td>
<td>Phenotype - value of phenotype (binary of continuous) - (0 or -9 if missing). <strong>Note:</strong> by default PLINK uses <code>2</code> and <code>1</code> for cases and controls, respectively, when processing a binary phenotype. However, you will see later that we can use <code>1</code> and <code>0</code> for cases and controls by providing the <code>--1</code> command line option when asking PLINK to handle a binary phenotypes.</td>
</tr>
</tbody>
</table>
</section>
<section id="bim-file" class="level3" data-number="7.1.2">
<h3 data-number="7.1.2" class="anchored" data-anchor-id="bim-file"><span class="header-section-number">7.1.2</span> .bim file</h3>
<p>The .bim file contains information about the genetic variants we have genotype data for. The following table describes the columns in the <code>.bim</code> file that do not have a header</p>
<table class="caption-top table">
<colgroup>
<col style="width: 13%">
<col style="width: 86%">
</colgroup>
<thead>
<tr class="header">
<th>Column #</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>1</td>
<td>Chromosome the genetic variant is located</td>
</tr>
<tr class="even">
<td>2</td>
<td>Unique identifier for genetic variant (usually an rsid or coordinate based ID)</td>
</tr>
<tr class="odd">
<td>3</td>
<td>Genetic distance defined in centimorgans - often set to 0 and ignored)</td>
</tr>
<tr class="even">
<td>4</td>
<td>Base-pair position on chromosome</td>
</tr>
<tr class="odd">
<td>5</td>
<td>Allele 1 - the minor-allele by default. Numerical genotypes are coded to this allele by default</td>
</tr>
<tr class="even">
<td>6</td>
<td>Allele 2 - the major-allele</td>
</tr>
</tbody>
</table>
</section>
<section id="bed-file" class="level3" data-number="7.1.3">
<h3 data-number="7.1.3" class="anchored" data-anchor-id="bed-file"><span class="header-section-number">7.1.3</span> .bed file</h3>
<p><code>.bed</code> files contain bit pairs that represent genotypes (e.g.&nbsp;<code>00</code> = homozygous for allele 1, <code>10</code> = heterozygous, <code>11</code> = homozygous for allele 2, <code>01</code> = missing.</p>
<p>The first 3 bytes of the .bed file are allocated and set for programs that analyse genetic data to recognized the format of the file. The first 2-bytes of the .bed file make up a <code>magic number</code> - a binary pattern that enables some programs that perform genetic analyses to identify this file as a file that contains genotype data coded as bit-pairs. The 3rd byte defines the <code>mode</code> that details how the genotypes are ordered in <code>.bed</code> file. The default mode is for blocks of bytes to be ordered by variants, and bits are ordered by individual (as opposed to blocks of bytes being ordered by individuals, and bit-pairs ordered by variants). The first 4 individuals will appear in the first byte associated with genotypes, with bit-pairs read from right to left (Least Significant Bit -&gt; Most Significant Bit) - i.e.&nbsp;the bit-pairs are ordered from right to left.</p>
<pre><code>| Magic Number  |  | Mode |  | Genotypes...       |
01101100 00011011  00000001  01101011 11011100  ...</code></pre>
</section>
</section>
<section id="use-plink-to-read-the-files-and-calculate-allele-frequencies" class="level2" data-number="7.2">
<h2 data-number="7.2" class="anchored" data-anchor-id="use-plink-to-read-the-files-and-calculate-allele-frequencies"><span class="header-section-number">7.2</span> Use PLINK to read the files and calculate allele frequencies</h2>
<p>We are going to use PLINK to read these files and calculate the allele frequencies to check our previous calculations were correct. We can do this by supplying PLINK the file name using <code>--bfile</code>, telling PLINK we want to calculate the allele frequencies of each variant using <code>--freq</code>, and instructing PLINK to output the data into a file with a specific prefix using <code>--out</code>:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb14"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="ex">/home/ubuntu/hpdm208z/programs/plink</span> <span class="dt">\</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">--bfile</span> workshop <span class="dt">\</span></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">--freq</span> <span class="dt">\</span></span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">--out</span> workshop_plink_freq_calcs</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>On successfully executing the command, you should see something similar to the following:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb15"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="ex">PLINK</span> v1.9.0-b.7.11 64-bit <span class="er">(</span><span class="ex">19</span> Aug 2025<span class="kw">)</span>           <span class="ex">cog-genomics.org/plink/1.9/</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="kw">(</span><span class="ex">C</span><span class="kw">)</span> <span class="ex">2005-2025</span> Shaun Purcell, Christopher Chang   GNU General Public License v3</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a><span class="ex">Logging</span> to workshop_plink_freq_calcs.log.</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a><span class="ex">Options</span> in effect:</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>  <span class="ex">--bfile</span> workshop</span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>  <span class="ex">--freq</span></span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a>  <span class="ex">--out</span> workshop_plink_freq_calcs</span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a><span class="ex">15986</span> MB RAM detected<span class="kw">;</span> <span class="ex">reserving</span> 7993 MB for main workspace.</span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a><span class="ex">5</span> variants loaded from .bim file.</span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a><span class="ex">10</span> people <span class="er">(</span><span class="ex">0</span> males, 0 females, 10 ambiguous<span class="kw">)</span> <span class="ex">loaded</span> from .fam.</span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a><span class="ex">Ambiguous</span> sex IDs written to workshop_plink_freq_calcs.nosex .</span>
<span id="cb15-13"><a href="#cb15-13" aria-hidden="true" tabindex="-1"></a><span class="ex">Using</span> 1 thread <span class="er">(</span><span class="ex">no</span> multithreaded calculations invoked<span class="kw">)</span><span class="bu">.</span></span>
<span id="cb15-14"><a href="#cb15-14" aria-hidden="true" tabindex="-1"></a><span class="ex">Before</span> main variant filters, 10 founders and 0 nonfounders present.</span>
<span id="cb15-15"><a href="#cb15-15" aria-hidden="true" tabindex="-1"></a><span class="ex">Calculating</span> allele frequencies... done.</span>
<span id="cb15-16"><a href="#cb15-16" aria-hidden="true" tabindex="-1"></a><span class="ex">Total</span> genotyping rate is exactly 1.</span>
<span id="cb15-17"><a href="#cb15-17" aria-hidden="true" tabindex="-1"></a><span class="ex">--freq:</span> Allele frequencies <span class="er">(</span><span class="ex">founders</span> only<span class="kw">)</span> <span class="ex">written</span> to</span>
<span id="cb15-18"><a href="#cb15-18" aria-hidden="true" tabindex="-1"></a><span class="ex">workshop_plink_freq_calcs.frq</span> .</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>The above output provides a summary of the options provided to PLINK, number of variants and individuals to process genotype data for, and the file names of any outputs generated. PLINK has informed us that it has output the calculated allele frequencies to <code>workshop_plink_freq_calcs.frq</code>.</p>
<p>Let’s take a look at the file using <code>more</code> on the Linux command line:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb16"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">more</span> workshop_plink_freq_calcs.frq</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>The contents of the file returned is:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb17"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a> <span class="ex">CHR</span>        SNP   A1   A2          MAF  NCHROBS</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>   <span class="ex">1</span>   rs543874    G    A          0.4       20</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>   <span class="ex">2</span> rs13021737    A    G          0.3       20</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>   <span class="ex">4</span> rs10938397    G    A          0.2       20</span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>  <span class="ex">16</span>  rs1558902    A    T          0.3       20</span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>  <span class="ex">18</span>  rs6567160    C    T          0.4       20</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>The first 4 columns contain informatiomn about the chromosome, identifier of the genetic variant and the 2 alleles specified in the .fam file (by default A1 will be the minor allele). The minor allele frequency has been returned based on the genotypes in the file in the column <code>MAF</code>. In addition, a column labelled <code>NCHROBS</code> has been returned. This is the total number of alleles observed for each variant. As genotypes of a given variant for an individual are composed of 2-alleles, then the number of alleles = number of individuals x 2.</p>
<section id="task" class="level3" data-number="7.2.1">
<h3 data-number="7.2.1" class="anchored" data-anchor-id="task"><span class="header-section-number">7.2.1</span> Task</h3>
<ul>
<li>Check the minor allele frequencies with your previous calculations above.</li>
</ul>
</section>
</section>
<section id="use-plink-to-convert-the-non-human-readble-genotype-data-to-human-readable-format-for-analysis" class="level2" data-number="7.3">
<h2 data-number="7.3" class="anchored" data-anchor-id="use-plink-to-convert-the-non-human-readble-genotype-data-to-human-readable-format-for-analysis"><span class="header-section-number">7.3</span> Use PLINK to convert the non-human readble genotype data to human-readable format for analysis</h2>
<p>A useful function of PLINK is its ability to convert genetic data to different formats. If we wanted to analyse the genetic data stored in the .bed .bim .fam files for analysis in Python or R, this could be problematic given the format of the data.</p>
<p>We will use PLINK to combine the data across the 3 files and export the data into a single file that can be easily loaded into a data frame. We will use the <code>--bfile</code>, <code>--recodeA</code> and <code>--out</code> command line options to do this:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb18"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="ex">/home/ubuntu/hpdm208z/programs/plink</span> <span class="dt">\</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">--bfile</span> workshop <span class="dt">\</span></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">--recodeA</span> <span class="dt">\</span></span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">--out</span> workshop_for_data_frames</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>The file output from this command we want to explore is <code>workshop_for_data_frames.raw</code>. We can look at this in the console using either <code>more</code> or <code>cat</code>.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb19"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">more</span> workshop_for_data_frames.raw</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>or</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb20"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span> workshop_for_data_frames.raw</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>Output:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb21"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="ex">FID</span> IID PAT MAT SEX PHENOTYPE rs543874_G rs13021737_A rs10938397_G rs1558902_A rs6567160_C</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="ex">A</span> A 0 0 0 <span class="at">-9</span> 2 2 1 1 0</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a><span class="ex">B</span> B 0 0 0 <span class="at">-9</span> 1 0 0 0 1</span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a><span class="ex">C</span> C 0 0 0 <span class="at">-9</span> 0 1 2 1 2</span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a><span class="ex">D</span> D 0 0 0 <span class="at">-9</span> 2 0 0 1 1</span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a><span class="ex">E</span> E 0 0 0 <span class="at">-9</span> 0 1 0 0 1</span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a><span class="ex">F</span> F 0 0 0 <span class="at">-9</span> 0 1 0 1 2</span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a><span class="ex">G</span> G 0 0 0 <span class="at">-9</span> 1 0 0 1 1</span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a><span class="ex">H</span> H 0 0 0 <span class="at">-9</span> 0 0 0 1 0</span>
<span id="cb21-10"><a href="#cb21-10" aria-hidden="true" tabindex="-1"></a><span class="ex">I</span> I 0 0 0 <span class="at">-9</span> 2 0 1 0 0</span>
<span id="cb21-11"><a href="#cb21-11" aria-hidden="true" tabindex="-1"></a><span class="ex">J</span> J 0 0 0 <span class="at">-9</span> 0 1 0 0 0</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>In this file, the first 6 columns are those found in the <code>.fam</code> file. The remaining columns represent the numerical genotypes of the genetic variants as ordered in the <code>.bim</code> file, with the allele the genotype is coded to appended to the unique name of the genetic variant specified in the <code>.bim</code> file. By default this will be the minor allele.</p>
<p>The genotypes match the numerical representation generated in section 6.2 where the genotypes were already coded to the minor allele.</p>
<p>Because the file extension is <code>.raw</code>, the <code>less</code> command might complain as it’s expecting .raw files to have a specific fie specification. You would need to change the name the file to something like <code>.txt</code> for <code>less</code> not to complain.</p>
</section>
<section id="subsetting-variants-and-individuals" class="level2" data-number="7.4">
<h2 data-number="7.4" class="anchored" data-anchor-id="subsetting-variants-and-individuals"><span class="header-section-number">7.4</span> Subsetting variants and individuals</h2>
<p>We can also use PLINK to subset genetic variants and/or individuals. If we wanted to only extract genotype data for SNPs <code>rs543874</code> and <code>rs13021737</code>, and output to a human readable format, we can do that by adding in the <code>--extract</code> option on the command line followed by the name of file that lists these SNP IDs on each line.</p>
<p>Copy the following text and save to a file within the <code>~/hpdm208z/workshops/genetic_data/</code> directory called <code>snps_to_extract.txt</code> :</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb22"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="ex">rs543874</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a><span class="ex">rs13021737</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>Similarly we can use PLINK to extract a subset of individuals we want to extract genotype data for by adding the <code>--keep</code> option on the command line followed by the name of a file that lists the individuals we want to keep on each line of the file. This file must contain the family ID (FID)) and the individual ID (IID) values found in the <code>.fam</code> file.</p>
<p>Copy the following text and save to a file within the <code>~/hpdm208z/workshops/genetic_data/</code> directory called <code>individuals_to_extract.txt</code> :</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb23"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="ex">A</span> A</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a><span class="ex">E</span> E</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a><span class="ex">J</span> J</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<section id="task-1" class="level3" data-number="7.4.1">
<h3 data-number="7.4.1" class="anchored" data-anchor-id="task-1"><span class="header-section-number">7.4.1</span> Task</h3>
<ul>
<li>Attempt to extract only the SNPs in listed in the <code>snps_to_extract.txt</code> file when converting the data to human-readable format.</li>
<li>Attempt to extract only the individuals in listed in the <code>individuals_to_extract.txt</code> file when converting to human-readable format.</li>
<li>Attempt to extract only the SNPs and individuals listed in both files when converting to human-readable format.</li>
</ul>
</section>
</section>
<section id="changing-the-genotype-codes-to-represent-numbers-of-a-specific-allele" class="level2" data-number="7.5">
<h2 data-number="7.5" class="anchored" data-anchor-id="changing-the-genotype-codes-to-represent-numbers-of-a-specific-allele"><span class="header-section-number">7.5</span> Changing the genotype codes to represent numbers of a specific allele</h2>
<p>As already mentioned, genotypes are coded to the number of copies of allele 1 (or <code>A1</code>) defined in the .bim file which by default is the minor allele. However, we may want to change the representaion of the genotypes so that they are coded to the major allele. This can be useful when we need to align genotype data with other datasets (we will look at a specific example of this next week).</p>
<p>We can use PLINK to do this by providing PLINK the option <code>--a1-allele</code> on the command line followed by a list of unique SNP identifiers and the allele we want the genotypes to be coded to.</p>
<p>Copy the following text and save to a file within the <code>~/hpdm208z/workshops/genetic_data/</code> directory called <code>reference_alleles.txt</code> :</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb24"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="ex">rs543874</span> A</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a><span class="ex">rs13021737</span> G</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>Now we can ask PLINK to change the genotype coding of the first two variants in the genotype data. As a reminder, here is the original numerical genotype file that was output:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb25"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="ex">FID</span> IID PAT MAT SEX PHENOTYPE rs543874_G rs13021737_A rs10938397_G rs1558902_A rs6567160_C</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a><span class="ex">A</span> A 0 0 0 <span class="at">-9</span> 2 2 1 1 0</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a><span class="ex">B</span> B 0 0 0 <span class="at">-9</span> 1 0 0 0 1</span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a><span class="ex">C</span> C 0 0 0 <span class="at">-9</span> 0 1 2 1 2</span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a><span class="ex">D</span> D 0 0 0 <span class="at">-9</span> 2 0 0 1 1</span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a><span class="ex">E</span> E 0 0 0 <span class="at">-9</span> 0 1 0 0 1</span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a><span class="ex">F</span> F 0 0 0 <span class="at">-9</span> 0 1 0 1 2</span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a><span class="ex">G</span> G 0 0 0 <span class="at">-9</span> 1 0 0 1 1</span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a><span class="ex">H</span> H 0 0 0 <span class="at">-9</span> 0 0 0 1 0</span>
<span id="cb25-10"><a href="#cb25-10" aria-hidden="true" tabindex="-1"></a><span class="ex">I</span> I 0 0 0 <span class="at">-9</span> 2 0 1 0 0</span>
<span id="cb25-11"><a href="#cb25-11" aria-hidden="true" tabindex="-1"></a><span class="ex">J</span> J 0 0 0 <span class="at">-9</span> 0 1 0 0 0</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>Now we will ask PLINK to modify the genotype codes:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb26"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="ex">/home/ubuntu/hpdm208z/programs/plink</span> <span class="dt">\</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">--bfile</span> workshop <span class="dt">\</span></span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">--a1-allele</span> reference_alleles.txt <span class="dt">\</span></span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">--recodeA</span> <span class="dt">\</span></span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">--out</span> workshop_for_data_frames_a1_changes</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>This command will create a file called <code>workshop_for_data_frames_a1_changes.raw</code> with the following contents:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb27"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="ex">FID</span> IID PAT MAT SEX PHENOTYPE rs543874_A rs13021737_G rs10938397_G rs1558902_A rs6567160_C</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a><span class="ex">A</span> A 0 0 0 <span class="at">-9</span> 0 0 1 1 0</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a><span class="ex">B</span> B 0 0 0 <span class="at">-9</span> 1 2 0 0 1</span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a><span class="ex">C</span> C 0 0 0 <span class="at">-9</span> 2 1 2 1 2</span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a><span class="ex">D</span> D 0 0 0 <span class="at">-9</span> 0 2 0 1 1</span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a><span class="ex">E</span> E 0 0 0 <span class="at">-9</span> 2 1 0 0 1</span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a><span class="ex">F</span> F 0 0 0 <span class="at">-9</span> 2 1 0 1 2</span>
<span id="cb27-8"><a href="#cb27-8" aria-hidden="true" tabindex="-1"></a><span class="ex">G</span> G 0 0 0 <span class="at">-9</span> 1 2 0 1 1</span>
<span id="cb27-9"><a href="#cb27-9" aria-hidden="true" tabindex="-1"></a><span class="ex">H</span> H 0 0 0 <span class="at">-9</span> 2 2 0 1 0</span>
<span id="cb27-10"><a href="#cb27-10" aria-hidden="true" tabindex="-1"></a><span class="ex">I</span> I 0 0 0 <span class="at">-9</span> 0 2 1 0 0</span>
<span id="cb27-11"><a href="#cb27-11" aria-hidden="true" tabindex="-1"></a><span class="ex">J</span> J 0 0 0 <span class="at">-9</span> 2 1 0 0 0</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>We can see that the genotype values of the first two SNPs in columns 7 and 8 of the file have now changed to reflect the number of copies of the <code>A</code> and <code>G</code> allele carried for each variant respectively for each individual.</p>
<p>For example, the fourth individual in the file with <code>IID = D</code> originally had a genotype code of <code>2</code> for <code>rs543874</code> when the genotypes for this SNP were coded to <code>G</code> to indicate they carry two copies of the <code>G</code> allele at the SNP site. That means this individual must not be carrying any copies of the A allele and this is reflected when changing from G to A as the allele the genotypes are coded to for the first variant. Values of 1 for other individuals are not affected because both the A and G are carried and so changing the alleles the genotypes correspond to has no affect.</p>
</section>
<section id="task-2" class="level2" data-number="7.6">
<h2 data-number="7.6" class="anchored" data-anchor-id="task-2"><span class="header-section-number">7.6</span> Task</h2>
<ul>
<li>look at the PLINK <a href="https://www.cog-genomics.org/plink/1.9/filter">documentation</a> to filter variants out of the data set with a minor allele frequency &lt; 0.4 when exporting to human-readble format.</li>
</ul>
</section>
<section id="running-a-genetic-association-analysis" class="level2" data-number="7.7">
<h2 data-number="7.7" class="anchored" data-anchor-id="running-a-genetic-association-analysis"><span class="header-section-number">7.7</span> Running a genetic association analysis</h2>
<p>In the final part of the workshop, we are going to ask PLINK to run a type-2 diabetes (T2D) genetic association analysis for us.</p>
<p>In the <code>t2d_study_5k_subs_10k_vars</code> file set, there are 1,000 cases and 4,000 controls for T2D, labelled <code>2</code> and <code>1</code> in the <code>.fam</code>, respectively. We will test the association for the 10k genetic variants to see if any of the genetic variants are associated with T2D status in our dataset.</p>
<p>Run the following command:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb28"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="ex">~/hpdm08z/programs/plink</span> <span class="dt">\</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">--bfile</span> t2d_study_5k <span class="dt">\</span></span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">--logistic</span> <span class="dt">\</span></span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">--out</span> t2d_gwas</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>Here is a table that describes the columns in the output</p>
<table class="caption-top table">
<colgroup>
<col style="width: 22%">
<col style="width: 77%">
</colgroup>
<thead>
<tr class="header">
<th>Column</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>CHR</code></td>
<td>From .bim file - chromosome the genetic variant is located</td>
</tr>
<tr class="even">
<td><code>SNP</code></td>
<td>From .bim file - unique name of genetic variant</td>
</tr>
<tr class="odd">
<td><code>BP</code></td>
<td>From .bim file - base-pair position genetic variant found</td>
</tr>
<tr class="even">
<td><code>A1</code></td>
<td>From .bim file - allele assigned A1 (usually minor allele) the BETA corresponds to</td>
</tr>
<tr class="odd">
<td><code>TEST</code></td>
<td>Describes model assumption - <code>ADD</code> = additive model of association</td>
</tr>
<tr class="even">
<td><code>NMISS</code></td>
<td>Number of individuals that went into t allele on the outcome</td>
</tr>
<tr class="odd">
<td><code>STAT</code></td>
<td>Z-statistic calculated as BETA / standard error (not shown) used to obtain P-value</td>
</tr>
<tr class="even">
<td><code>P</code></td>
<td>P-value indicating strength of statistical significance of estimated effect</td>
</tr>
</tbody>
</table>
<section id="questions" class="level3" data-number="7.7.1">
<h3 data-number="7.7.1" class="anchored" data-anchor-id="questions"><span class="header-section-number">7.7.1</span> Questions</h3>
<ul>
<li>Are there any genome-wide significant associations reaching P&lt;5X10-8?</li>
<li>For the genetic variant with the smallest P-value, what does the odds ratio tell us?</li>
</ul>
<p>We are going to look more at genetic association testing next week, including meta-analysis and potential confounding of statistical associations.</p>
</section>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
      const outerScaffold = trigger.parentElement.cloneNode(true);
      const codeEl = outerScaffold.querySelector('code');
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




</body></html>